{
 "metadata": {
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.0"
  },
  "orig_nbformat": 2,
  "kernelspec": {
   "name": "python3",
   "display_name": "Python 3.9.0 64-bit"
  },
  "interpreter": {
   "hash": "50c1d9dc61b70a857575e6cd78e3f7f868bc672eff5937ba251fee1ba26d9a38"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2,
 "cells": [
  {
   "source": [
    "# Table of Contents\n",
    "1) Sequence Manipulation\n",
    "\n",
    "2) Protein Synthesis\n",
    "\n",
    "3) DNA Composition - GC and AT content\n",
    "\n",
    "4) Sequence Alignment\n",
    "\n",
    "5) Difference between Sequence Similarity and Sequence Identity\n",
    "\n",
    "6) Checking Similarity Using Distance\n",
    "\n",
    "7) Checking for Similarity using DotPlot\n",
    "\n",
    "8) Working with Biological Data File Formats\n",
    "\n",
    "9) Working with 3D structures\n",
    "\n",
    "10) Visualizing 3D structures\n",
    "\n",
    "11) BLAST\n"
   ],
   "cell_type": "markdown",
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import Bio"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Seq import Seq\n",
    "#dir(Seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "seq1 = (\"ATGATCTCGTAA\")\n",
    "\n",
    "# Create a simple DNA sequence\n",
    "dna = Seq(seq1)\n",
    "\n",
    "#Create a simple RNA sequence\n",
    "rna = Seq(seq1.replace('T','U'))\n",
    "\n",
    "#Create a simple protein\n",
    "protein_a = Seq(\"MIT\")"
   ]
  },
  {
   "source": [
    "# 1) Sequence Manipulation\n",
    "\n"
   ],
   "cell_type": "markdown",
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "2"
      ]
     },
     "metadata": {},
     "execution_count": 41
    }
   ],
   "source": [
    "#Length of sequence\n",
    "len(dna)\n",
    "\n",
    "#Slicing to obtain first 3 nucleotides\n",
    "dna[0:3]\n",
    "\n",
    "#Reverse a sequence\n",
    "dna[::-1]\n",
    "\n",
    "#Combining sequences\n",
    "seq2 = Seq(\"ATGATCTCGTGG\")\n",
    "dna[0:6] + seq2 # ATGATCAATGATCTCGTAA - adds first 6 nucleotides of dna to seq 2 \n",
    "\n",
    "#Find number of a particular nucleotide within sequence\n",
    "dna.count(\"G\") #4\n",
    "\n",
    "#Find number of codons occurring in a particular sequence(3N)\n",
    "dna.count(\"ATG\") #1\n",
    "\n",
    "#Find first position/location of a nucleotide using FIND from LEFT\n",
    "#Recall python 0 based indexing\n",
    "dna.find(\"G\") #2\n",
    "\n",
    "#First first position/location of a nucleotide using FIND from RIGHT\n",
    "dna.rfind(\"G\") #8\n",
    "\n",
    "#Find location/index of a nucleotide using INDEX\n",
    "dna.index(\"G\") #2 \n",
    "\n"
   ]
  },
  {
   "source": [
    "### Plot of frequency of nucleotides\n"
   ],
   "cell_type": "markdown",
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "from collections import Counter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "<BarContainer object of 4 artists>"
      ]
     },
     "metadata": {},
     "execution_count": 44
    },
    {
     "output_type": "display_data",
     "data": {
      "text/plain": "<Figure size 432x288 with 1 Axes>",
      "image/svg+xml": "<?xml version=\"1.0\" encoding=\"utf-8\" standalone=\"no\"?>\r\n<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\r\n  \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\r\n<svg height=\"248.518125pt\" version=\"1.1\" viewBox=\"0 0 372.103125 248.518125\" width=\"372.103125pt\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\r\n <metadata>\r\n  <rdf:RDF xmlns:cc=\"http://creativecommons.org/ns#\" xmlns:dc=\"http://purl.org/dc/elements/1.1/\" xmlns:rdf=\"http://www.w3.org/1999/02/22-rdf-syntax-ns#\">\r\n   <cc:Work>\r\n    <dc:type rdf:resource=\"http://purl.org/dc/dcmitype/StillImage\"/>\r\n    <dc:date>2021-04-17T10:27:54.911044</dc:date>\r\n    <dc:format>image/svg+xml</dc:format>\r\n    <dc:creator>\r\n     <cc:Agent>\r\n      <dc:title>Matplotlib v3.4.0, https://matplotlib.org/</dc:title>\r\n     </cc:Agent>\r\n    </dc:creator>\r\n   </cc:Work>\r\n  </rdf:RDF>\r\n </metadata>\r\n <defs>\r\n  <style type=\"text/css\">*{stroke-linecap:butt;stroke-linejoin:round;}</style>\r\n </defs>\r\n <g id=\"figure_1\">\r\n  <g id=\"patch_1\">\r\n   <path d=\"M 0 248.518125 \r\nL 372.103125 248.518125 \r\nL 372.103125 0 \r\nL 0 0 \r\nz\r\n\" style=\"fill:none;\"/>\r\n  </g>\r\n  <g id=\"axes_1\">\r\n   <g id=\"patch_2\">\r\n    <path d=\"M 30.103125 224.64 \r\nL 364.903125 224.64 \r\nL 364.903125 7.2 \r\nL 30.103125 7.2 \r\nz\r\n\" style=\"fill:#ffffff;\"/>\r\n   </g>\r\n   <g id=\"patch_3\">\r\n    <path clip-path=\"url(#pdb4c489d1b)\" d=\"M 45.321307 224.64 \r\nL 109.397862 224.64 \r\nL 109.397862 17.554286 \r\nL 45.321307 17.554286 \r\nz\r\n\" style=\"fill:#1f77b4;\"/>\r\n   </g>\r\n   <g id=\"patch_4\">\r\n    <path clip-path=\"url(#pdb4c489d1b)\" d=\"M 125.417001 224.64 \r\nL 189.493556 224.64 \r\nL 189.493556 17.554286 \r\nL 125.417001 17.554286 \r\nz\r\n\" style=\"fill:#1f77b4;\"/>\r\n   </g>\r\n   <g id=\"patch_5\">\r\n    <path clip-path=\"url(#pdb4c489d1b)\" d=\"M 205.512694 224.64 \r\nL 269.589249 224.64 \r\nL 269.589249 121.097143 \r\nL 205.512694 121.097143 \r\nz\r\n\" style=\"fill:#1f77b4;\"/>\r\n   </g>\r\n   <g id=\"patch_6\">\r\n    <path clip-path=\"url(#pdb4c489d1b)\" d=\"M 285.608388 224.64 \r\nL 349.684943 224.64 \r\nL 349.684943 121.097143 \r\nL 285.608388 121.097143 \r\nz\r\n\" style=\"fill:#1f77b4;\"/>\r\n   </g>\r\n   <g id=\"matplotlib.axis_1\">\r\n    <g id=\"xtick_1\">\r\n     <g id=\"line2d_1\">\r\n      <defs>\r\n       <path d=\"M 0 0 \r\nL 0 3.5 \r\n\" id=\"m070a596fd4\" style=\"stroke:#000000;stroke-width:0.8;\"/>\r\n      </defs>\r\n      <g>\r\n       <use style=\"stroke:#000000;stroke-width:0.8;\" x=\"77.359584\" xlink:href=\"#m070a596fd4\" y=\"224.64\"/>\r\n      </g>\r\n     </g>\r\n     <g id=\"text_1\">\r\n      <!-- A -->\r\n      <g transform=\"translate(73.939272 239.238437)scale(0.1 -0.1)\">\r\n       <defs>\r\n        <path d=\"M 2188 4044 \r\nL 1331 1722 \r\nL 3047 1722 \r\nL 2188 4044 \r\nz\r\nM 1831 4666 \r\nL 2547 4666 \r\nL 4325 0 \r\nL 3669 0 \r\nL 3244 1197 \r\nL 1141 1197 \r\nL 716 0 \r\nL 50 0 \r\nL 1831 4666 \r\nz\r\n\" id=\"DejaVuSans-41\" transform=\"scale(0.015625)\"/>\r\n       </defs>\r\n       <use xlink:href=\"#DejaVuSans-41\"/>\r\n      </g>\r\n     </g>\r\n    </g>\r\n    <g id=\"xtick_2\">\r\n     <g id=\"line2d_2\">\r\n      <g>\r\n       <use style=\"stroke:#000000;stroke-width:0.8;\" x=\"157.455278\" xlink:href=\"#m070a596fd4\" y=\"224.64\"/>\r\n      </g>\r\n     </g>\r\n     <g id=\"text_2\">\r\n      <!-- T -->\r\n      <g transform=\"translate(154.401372 239.238437)scale(0.1 -0.1)\">\r\n       <defs>\r\n        <path d=\"M -19 4666 \r\nL 3928 4666 \r\nL 3928 4134 \r\nL 2272 4134 \r\nL 2272 0 \r\nL 1638 0 \r\nL 1638 4134 \r\nL -19 4134 \r\nL -19 4666 \r\nz\r\n\" id=\"DejaVuSans-54\" transform=\"scale(0.015625)\"/>\r\n       </defs>\r\n       <use xlink:href=\"#DejaVuSans-54\"/>\r\n      </g>\r\n     </g>\r\n    </g>\r\n    <g id=\"xtick_3\">\r\n     <g id=\"line2d_3\">\r\n      <g>\r\n       <use style=\"stroke:#000000;stroke-width:0.8;\" x=\"237.550972\" xlink:href=\"#m070a596fd4\" y=\"224.64\"/>\r\n      </g>\r\n     </g>\r\n     <g id=\"text_3\">\r\n      <!-- G -->\r\n      <g transform=\"translate(233.676753 239.238437)scale(0.1 -0.1)\">\r\n       <defs>\r\n        <path d=\"M 3809 666 \r\nL 3809 1919 \r\nL 2778 1919 \r\nL 2778 2438 \r\nL 4434 2438 \r\nL 4434 434 \r\nQ 4069 175 3628 42 \r\nQ 3188 -91 2688 -91 \r\nQ 1594 -91 976 548 \r\nQ 359 1188 359 2328 \r\nQ 359 3472 976 4111 \r\nQ 1594 4750 2688 4750 \r\nQ 3144 4750 3555 4637 \r\nQ 3966 4525 4313 4306 \r\nL 4313 3634 \r\nQ 3963 3931 3569 4081 \r\nQ 3175 4231 2741 4231 \r\nQ 1884 4231 1454 3753 \r\nQ 1025 3275 1025 2328 \r\nQ 1025 1384 1454 906 \r\nQ 1884 428 2741 428 \r\nQ 3075 428 3337 486 \r\nQ 3600 544 3809 666 \r\nz\r\n\" id=\"DejaVuSans-47\" transform=\"scale(0.015625)\"/>\r\n       </defs>\r\n       <use xlink:href=\"#DejaVuSans-47\"/>\r\n      </g>\r\n     </g>\r\n    </g>\r\n    <g id=\"xtick_4\">\r\n     <g id=\"line2d_4\">\r\n      <g>\r\n       <use style=\"stroke:#000000;stroke-width:0.8;\" x=\"317.646666\" xlink:href=\"#m070a596fd4\" y=\"224.64\"/>\r\n      </g>\r\n     </g>\r\n     <g id=\"text_4\">\r\n      <!-- C -->\r\n      <g transform=\"translate(314.155259 239.238437)scale(0.1 -0.1)\">\r\n       <defs>\r\n        <path d=\"M 4122 4306 \r\nL 4122 3641 \r\nQ 3803 3938 3442 4084 \r\nQ 3081 4231 2675 4231 \r\nQ 1875 4231 1450 3742 \r\nQ 1025 3253 1025 2328 \r\nQ 1025 1406 1450 917 \r\nQ 1875 428 2675 428 \r\nQ 3081 428 3442 575 \r\nQ 3803 722 4122 1019 \r\nL 4122 359 \r\nQ 3791 134 3420 21 \r\nQ 3050 -91 2638 -91 \r\nQ 1578 -91 968 557 \r\nQ 359 1206 359 2328 \r\nQ 359 3453 968 4101 \r\nQ 1578 4750 2638 4750 \r\nQ 3056 4750 3426 4639 \r\nQ 3797 4528 4122 4306 \r\nz\r\n\" id=\"DejaVuSans-43\" transform=\"scale(0.015625)\"/>\r\n       </defs>\r\n       <use xlink:href=\"#DejaVuSans-43\"/>\r\n      </g>\r\n     </g>\r\n    </g>\r\n   </g>\r\n   <g id=\"matplotlib.axis_2\">\r\n    <g id=\"ytick_1\">\r\n     <g id=\"line2d_5\">\r\n      <defs>\r\n       <path d=\"M 0 0 \r\nL -3.5 0 \r\n\" id=\"mdcb6112c81\" style=\"stroke:#000000;stroke-width:0.8;\"/>\r\n      </defs>\r\n      <g>\r\n       <use style=\"stroke:#000000;stroke-width:0.8;\" x=\"30.103125\" xlink:href=\"#mdcb6112c81\" y=\"224.64\"/>\r\n      </g>\r\n     </g>\r\n     <g id=\"text_5\">\r\n      <!-- 0.0 -->\r\n      <g transform=\"translate(7.2 228.439219)scale(0.1 -0.1)\">\r\n       <defs>\r\n        <path d=\"M 2034 4250 \r\nQ 1547 4250 1301 3770 \r\nQ 1056 3291 1056 2328 \r\nQ 1056 1369 1301 889 \r\nQ 1547 409 2034 409 \r\nQ 2525 409 2770 889 \r\nQ 3016 1369 3016 2328 \r\nQ 3016 3291 2770 3770 \r\nQ 2525 4250 2034 4250 \r\nz\r\nM 2034 4750 \r\nQ 2819 4750 3233 4129 \r\nQ 3647 3509 3647 2328 \r\nQ 3647 1150 3233 529 \r\nQ 2819 -91 2034 -91 \r\nQ 1250 -91 836 529 \r\nQ 422 1150 422 2328 \r\nQ 422 3509 836 4129 \r\nQ 1250 4750 2034 4750 \r\nz\r\n\" id=\"DejaVuSans-30\" transform=\"scale(0.015625)\"/>\r\n        <path d=\"M 684 794 \r\nL 1344 794 \r\nL 1344 0 \r\nL 684 0 \r\nL 684 794 \r\nz\r\n\" id=\"DejaVuSans-2e\" transform=\"scale(0.015625)\"/>\r\n       </defs>\r\n       <use xlink:href=\"#DejaVuSans-30\"/>\r\n       <use x=\"63.623047\" xlink:href=\"#DejaVuSans-2e\"/>\r\n       <use x=\"95.410156\" xlink:href=\"#DejaVuSans-30\"/>\r\n      </g>\r\n     </g>\r\n    </g>\r\n    <g id=\"ytick_2\">\r\n     <g id=\"line2d_6\">\r\n      <g>\r\n       <use style=\"stroke:#000000;stroke-width:0.8;\" x=\"30.103125\" xlink:href=\"#mdcb6112c81\" y=\"198.754286\"/>\r\n      </g>\r\n     </g>\r\n     <g id=\"text_6\">\r\n      <!-- 0.5 -->\r\n      <g transform=\"translate(7.2 202.553504)scale(0.1 -0.1)\">\r\n       <defs>\r\n        <path d=\"M 691 4666 \r\nL 3169 4666 \r\nL 3169 4134 \r\nL 1269 4134 \r\nL 1269 2991 \r\nQ 1406 3038 1543 3061 \r\nQ 1681 3084 1819 3084 \r\nQ 2600 3084 3056 2656 \r\nQ 3513 2228 3513 1497 \r\nQ 3513 744 3044 326 \r\nQ 2575 -91 1722 -91 \r\nQ 1428 -91 1123 -41 \r\nQ 819 9 494 109 \r\nL 494 744 \r\nQ 775 591 1075 516 \r\nQ 1375 441 1709 441 \r\nQ 2250 441 2565 725 \r\nQ 2881 1009 2881 1497 \r\nQ 2881 1984 2565 2268 \r\nQ 2250 2553 1709 2553 \r\nQ 1456 2553 1204 2497 \r\nQ 953 2441 691 2322 \r\nL 691 4666 \r\nz\r\n\" id=\"DejaVuSans-35\" transform=\"scale(0.015625)\"/>\r\n       </defs>\r\n       <use xlink:href=\"#DejaVuSans-30\"/>\r\n       <use x=\"63.623047\" xlink:href=\"#DejaVuSans-2e\"/>\r\n       <use x=\"95.410156\" xlink:href=\"#DejaVuSans-35\"/>\r\n      </g>\r\n     </g>\r\n    </g>\r\n    <g id=\"ytick_3\">\r\n     <g id=\"line2d_7\">\r\n      <g>\r\n       <use style=\"stroke:#000000;stroke-width:0.8;\" x=\"30.103125\" xlink:href=\"#mdcb6112c81\" y=\"172.868571\"/>\r\n      </g>\r\n     </g>\r\n     <g id=\"text_7\">\r\n      <!-- 1.0 -->\r\n      <g transform=\"translate(7.2 176.66779)scale(0.1 -0.1)\">\r\n       <defs>\r\n        <path d=\"M 794 531 \r\nL 1825 531 \r\nL 1825 4091 \r\nL 703 3866 \r\nL 703 4441 \r\nL 1819 4666 \r\nL 2450 4666 \r\nL 2450 531 \r\nL 3481 531 \r\nL 3481 0 \r\nL 794 0 \r\nL 794 531 \r\nz\r\n\" id=\"DejaVuSans-31\" transform=\"scale(0.015625)\"/>\r\n       </defs>\r\n       <use xlink:href=\"#DejaVuSans-31\"/>\r\n       <use x=\"63.623047\" xlink:href=\"#DejaVuSans-2e\"/>\r\n       <use x=\"95.410156\" xlink:href=\"#DejaVuSans-30\"/>\r\n      </g>\r\n     </g>\r\n    </g>\r\n    <g id=\"ytick_4\">\r\n     <g id=\"line2d_8\">\r\n      <g>\r\n       <use style=\"stroke:#000000;stroke-width:0.8;\" x=\"30.103125\" xlink:href=\"#mdcb6112c81\" y=\"146.982857\"/>\r\n      </g>\r\n     </g>\r\n     <g id=\"text_8\">\r\n      <!-- 1.5 -->\r\n      <g transform=\"translate(7.2 150.782076)scale(0.1 -0.1)\">\r\n       <use xlink:href=\"#DejaVuSans-31\"/>\r\n       <use x=\"63.623047\" xlink:href=\"#DejaVuSans-2e\"/>\r\n       <use x=\"95.410156\" xlink:href=\"#DejaVuSans-35\"/>\r\n      </g>\r\n     </g>\r\n    </g>\r\n    <g id=\"ytick_5\">\r\n     <g id=\"line2d_9\">\r\n      <g>\r\n       <use style=\"stroke:#000000;stroke-width:0.8;\" x=\"30.103125\" xlink:href=\"#mdcb6112c81\" y=\"121.097143\"/>\r\n      </g>\r\n     </g>\r\n     <g id=\"text_9\">\r\n      <!-- 2.0 -->\r\n      <g transform=\"translate(7.2 124.896362)scale(0.1 -0.1)\">\r\n       <defs>\r\n        <path d=\"M 1228 531 \r\nL 3431 531 \r\nL 3431 0 \r\nL 469 0 \r\nL 469 531 \r\nQ 828 903 1448 1529 \r\nQ 2069 2156 2228 2338 \r\nQ 2531 2678 2651 2914 \r\nQ 2772 3150 2772 3378 \r\nQ 2772 3750 2511 3984 \r\nQ 2250 4219 1831 4219 \r\nQ 1534 4219 1204 4116 \r\nQ 875 4013 500 3803 \r\nL 500 4441 \r\nQ 881 4594 1212 4672 \r\nQ 1544 4750 1819 4750 \r\nQ 2544 4750 2975 4387 \r\nQ 3406 4025 3406 3419 \r\nQ 3406 3131 3298 2873 \r\nQ 3191 2616 2906 2266 \r\nQ 2828 2175 2409 1742 \r\nQ 1991 1309 1228 531 \r\nz\r\n\" id=\"DejaVuSans-32\" transform=\"scale(0.015625)\"/>\r\n       </defs>\r\n       <use xlink:href=\"#DejaVuSans-32\"/>\r\n       <use x=\"63.623047\" xlink:href=\"#DejaVuSans-2e\"/>\r\n       <use x=\"95.410156\" xlink:href=\"#DejaVuSans-30\"/>\r\n      </g>\r\n     </g>\r\n    </g>\r\n    <g id=\"ytick_6\">\r\n     <g id=\"line2d_10\">\r\n      <g>\r\n       <use style=\"stroke:#000000;stroke-width:0.8;\" x=\"30.103125\" xlink:href=\"#mdcb6112c81\" y=\"95.211429\"/>\r\n      </g>\r\n     </g>\r\n     <g id=\"text_10\">\r\n      <!-- 2.5 -->\r\n      <g transform=\"translate(7.2 99.010647)scale(0.1 -0.1)\">\r\n       <use xlink:href=\"#DejaVuSans-32\"/>\r\n       <use x=\"63.623047\" xlink:href=\"#DejaVuSans-2e\"/>\r\n       <use x=\"95.410156\" xlink:href=\"#DejaVuSans-35\"/>\r\n      </g>\r\n     </g>\r\n    </g>\r\n    <g id=\"ytick_7\">\r\n     <g id=\"line2d_11\">\r\n      <g>\r\n       <use style=\"stroke:#000000;stroke-width:0.8;\" x=\"30.103125\" xlink:href=\"#mdcb6112c81\" y=\"69.325714\"/>\r\n      </g>\r\n     </g>\r\n     <g id=\"text_11\">\r\n      <!-- 3.0 -->\r\n      <g transform=\"translate(7.2 73.124933)scale(0.1 -0.1)\">\r\n       <defs>\r\n        <path d=\"M 2597 2516 \r\nQ 3050 2419 3304 2112 \r\nQ 3559 1806 3559 1356 \r\nQ 3559 666 3084 287 \r\nQ 2609 -91 1734 -91 \r\nQ 1441 -91 1130 -33 \r\nQ 819 25 488 141 \r\nL 488 750 \r\nQ 750 597 1062 519 \r\nQ 1375 441 1716 441 \r\nQ 2309 441 2620 675 \r\nQ 2931 909 2931 1356 \r\nQ 2931 1769 2642 2001 \r\nQ 2353 2234 1838 2234 \r\nL 1294 2234 \r\nL 1294 2753 \r\nL 1863 2753 \r\nQ 2328 2753 2575 2939 \r\nQ 2822 3125 2822 3475 \r\nQ 2822 3834 2567 4026 \r\nQ 2313 4219 1838 4219 \r\nQ 1578 4219 1281 4162 \r\nQ 984 4106 628 3988 \r\nL 628 4550 \r\nQ 988 4650 1302 4700 \r\nQ 1616 4750 1894 4750 \r\nQ 2613 4750 3031 4423 \r\nQ 3450 4097 3450 3541 \r\nQ 3450 3153 3228 2886 \r\nQ 3006 2619 2597 2516 \r\nz\r\n\" id=\"DejaVuSans-33\" transform=\"scale(0.015625)\"/>\r\n       </defs>\r\n       <use xlink:href=\"#DejaVuSans-33\"/>\r\n       <use x=\"63.623047\" xlink:href=\"#DejaVuSans-2e\"/>\r\n       <use x=\"95.410156\" xlink:href=\"#DejaVuSans-30\"/>\r\n      </g>\r\n     </g>\r\n    </g>\r\n    <g id=\"ytick_8\">\r\n     <g id=\"line2d_12\">\r\n      <g>\r\n       <use style=\"stroke:#000000;stroke-width:0.8;\" x=\"30.103125\" xlink:href=\"#mdcb6112c81\" y=\"43.44\"/>\r\n      </g>\r\n     </g>\r\n     <g id=\"text_12\">\r\n      <!-- 3.5 -->\r\n      <g transform=\"translate(7.2 47.239219)scale(0.1 -0.1)\">\r\n       <use xlink:href=\"#DejaVuSans-33\"/>\r\n       <use x=\"63.623047\" xlink:href=\"#DejaVuSans-2e\"/>\r\n       <use x=\"95.410156\" xlink:href=\"#DejaVuSans-35\"/>\r\n      </g>\r\n     </g>\r\n    </g>\r\n    <g id=\"ytick_9\">\r\n     <g id=\"line2d_13\">\r\n      <g>\r\n       <use style=\"stroke:#000000;stroke-width:0.8;\" x=\"30.103125\" xlink:href=\"#mdcb6112c81\" y=\"17.554286\"/>\r\n      </g>\r\n     </g>\r\n     <g id=\"text_13\">\r\n      <!-- 4.0 -->\r\n      <g transform=\"translate(7.2 21.353504)scale(0.1 -0.1)\">\r\n       <defs>\r\n        <path d=\"M 2419 4116 \r\nL 825 1625 \r\nL 2419 1625 \r\nL 2419 4116 \r\nz\r\nM 2253 4666 \r\nL 3047 4666 \r\nL 3047 1625 \r\nL 3713 1625 \r\nL 3713 1100 \r\nL 3047 1100 \r\nL 3047 0 \r\nL 2419 0 \r\nL 2419 1100 \r\nL 313 1100 \r\nL 313 1709 \r\nL 2253 4666 \r\nz\r\n\" id=\"DejaVuSans-34\" transform=\"scale(0.015625)\"/>\r\n       </defs>\r\n       <use xlink:href=\"#DejaVuSans-34\"/>\r\n       <use x=\"63.623047\" xlink:href=\"#DejaVuSans-2e\"/>\r\n       <use x=\"95.410156\" xlink:href=\"#DejaVuSans-30\"/>\r\n      </g>\r\n     </g>\r\n    </g>\r\n   </g>\r\n   <g id=\"patch_7\">\r\n    <path d=\"M 30.103125 224.64 \r\nL 30.103125 7.2 \r\n\" style=\"fill:none;stroke:#000000;stroke-linecap:square;stroke-linejoin:miter;stroke-width:0.8;\"/>\r\n   </g>\r\n   <g id=\"patch_8\">\r\n    <path d=\"M 364.903125 224.64 \r\nL 364.903125 7.2 \r\n\" style=\"fill:none;stroke:#000000;stroke-linecap:square;stroke-linejoin:miter;stroke-width:0.8;\"/>\r\n   </g>\r\n   <g id=\"patch_9\">\r\n    <path d=\"M 30.103125 224.64 \r\nL 364.903125 224.64 \r\n\" style=\"fill:none;stroke:#000000;stroke-linecap:square;stroke-linejoin:miter;stroke-width:0.8;\"/>\r\n   </g>\r\n   <g id=\"patch_10\">\r\n    <path d=\"M 30.103125 7.2 \r\nL 364.903125 7.2 \r\n\" style=\"fill:none;stroke:#000000;stroke-linecap:square;stroke-linejoin:miter;stroke-width:0.8;\"/>\r\n   </g>\r\n  </g>\r\n </g>\r\n <defs>\r\n  <clipPath id=\"pdb4c489d1b\">\r\n   <rect height=\"217.44\" width=\"334.8\" x=\"30.103125\" y=\"7.2\"/>\r\n  </clipPath>\r\n </defs>\r\n</svg>\r\n",
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXQAAAD4CAYAAAD8Zh1EAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjQuMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8+yak3AAAACXBIWXMAAAsTAAALEwEAmpwYAAAOwUlEQVR4nO3dfYhld33H8fenmzVaFEOzg4Z9yNiaUtRqNENUbGkaK8QHsooREqgaiQyIQQVpm/hHxPxT/ccUjRgWE1xTayLxgVXXSiARDdQ1s9tNdLOmbH0gu0Qy7urGoMaufPvHnMj0du7eMzN35u78fL/gMOec3/ee882BfPbMmXPuSVUhSdr4/mjSDUiSxsNAl6RGGOiS1AgDXZIaYaBLUiPOmtSOt2zZUtPT05PavSRtSPv37/9ZVU0tNTaxQJ+enmZubm5Su5ekDSnJT4aNeclFkhphoEtSIwx0SWqEgS5JjTDQJakRBrokNaJ3oCfZlOQ/k3x1ibGzk9yZ5EiSfUmmx9qlJGmk5Zyhvxc4PGTsGuDnVfV84CbgI6ttTJK0PL0CPck24PXAp4aU7AR2d/N3Aa9OktW3J0nqq++Tov8C/CPwrCHjW4FHAKrqVJKTwLnAzxYXJZkFZgF27NixgnYXTF/3tRV/tgU//vDrV70Nj+Hqj6F0phl5hp7kDcBjVbV/tTurql1VNVNVM1NTS34VgSRphfpccnkVcHmSHwN3AJcm+deBmmPAdoAkZwHPBo6PsU9J0ggjA72qrq+qbVU1DVwJ3FNVfz9Qtgd4ezd/RVfjy0olaR2t+NsWk9wIzFXVHuBW4PYkR4ATLAS/JGkdLSvQq+qbwDe7+RsWrf8N8JZxNiZJWh6fFJWkRhjoktQIA12SGmGgS1IjDHRJaoSBLkmNMNAlqREGuiQ1wkCXpEYY6JLUCANdkhphoEtSIwx0SWqEgS5JjTDQJakRBrokNaLPS6KfnuS7SR5IcijJh5aouTrJfJKD3fTOtWlXkjRMnzcWPQlcWlVPJNkM3Jfk61X1nYG6O6vq2vG3KEnqY2Sgdy97fqJb3NxNvgBaks4wva6hJ9mU5CDwGHB3Ve1bouzNSR5McleS7eNsUpI0Wq9Ar6rfVdWFwDbg4iQvGij5CjBdVS8G7gZ2L7WdJLNJ5pLMzc/Pr6JtSdKgZd3lUlW/AO4FLhtYf7yqnuwWPwVcNOTzu6pqpqpmpqamVtCuJGmYPne5TCU5p5t/BvAa4AcDNectWrwcODzGHiVJPfS5y+U8YHeSTSz8A/D5qvpqkhuBuaraA7wnyeXAKeAEcPVaNSxJWlqfu1weBF66xPobFs1fD1w/3tYkScvhk6KS1AgDXZIaYaBLUiMMdElqhIEuSY0w0CWpEQa6JDXCQJekRhjoktQIA12SGmGgS1IjDHRJaoSBLkmNMNAlqREGuiQ1wkCXpEYY6JLUiD7vFH16ku8meSDJoSQfWqLm7CR3JjmSZF+S6TXpVpI0VJ8z9CeBS6vqJcCFwGVJXjFQcw3w86p6PnAT8JGxdilJGmlkoNeCJ7rFzd1UA2U7gd3d/F3Aq5NkbF1KkkbqdQ09yaYkB4HHgLurat9AyVbgEYCqOgWcBM5dYjuzSeaSzM3Pz6+qcUnS/9Ur0Kvqd1V1IbANuDjJi1ays6raVVUzVTUzNTW1kk1IkoZY1l0uVfUL4F7gsoGhY8B2gCRnAc8Gjo+hP0lST33ucplKck43/wzgNcAPBsr2AG/v5q8A7qmqwevskqQ1dFaPmvOA3Uk2sfAPwOer6qtJbgTmqmoPcCtwe5IjwAngyjXrWJK0pJGBXlUPAi9dYv0Ni+Z/A7xlvK1JkpbDJ0UlqREGuiQ1wkCXpEYY6JLUCANdkhphoEtSIwx0SWqEgS5JjTDQJakRBrokNcJAl6RGGOiS1AgDXZIaYaBLUiMMdElqhIEuSY0w0CWpEX3eKbo9yb1JHkpyKMl7l6i5JMnJJAe76YaltiVJWjt93il6Cnh/VR1I8ixgf5K7q+qhgbpvV9Ubxt+iJKmPkWfoVfVoVR3o5n8JHAa2rnVjkqTlWdY19CTTLLwwet8Sw69M8kCSryd54ZDPzyaZSzI3Pz+//G4lSUP1DvQkzwS+ALyvqh4fGD4AnF9VLwE+Dnx5qW1U1a6qmqmqmampqRW2LElaSq9AT7KZhTD/bFV9cXC8qh6vqie6+b3A5iRbxtqpJOm0+tzlEuBW4HBVfXRIzXO7OpJc3G33+DgblSSdXp+7XF4FvBX4XpKD3boPADsAquoW4ArgXUlOAb8GrqyqGn+7kqRhRgZ6Vd0HZETNzcDN42pKkrR8PikqSY0w0CWpEQa6JDXCQJekRhjoktQIA12SGmGgS1IjDHRJaoSBLkmNMNAlqREGuiQ1wkCXpEYY6JLUCANdkhphoEtSIwx0SWqEgS5JjejzTtHtSe5N8lCSQ0neu0RNknwsyZEkDyZ52dq0K0kaps87RU8B76+qA0meBexPcndVPbSo5rXABd30cuCT3U9J0joZeYZeVY9W1YFu/pfAYWDrQNlO4DO14DvAOUnOG3u3kqSh+pyh/16SaeClwL6Boa3AI4uWj3brHh34/CwwC7Bjx45ltiqdOaav+9qkW5ioH3/49av6vMdvdcdvmN5/FE3yTOALwPuq6vGV7KyqdlXVTFXNTE1NrWQTkqQhegV6ks0shPlnq+qLS5QcA7YvWt7WrZMkrZM+d7kEuBU4XFUfHVK2B3hbd7fLK4CTVfXokFpJ0hrocw39VcBbge8lOdit+wCwA6CqbgH2Aq8DjgC/At4x9k4lSac1MtCr6j4gI2oKePe4mpIkLZ9PikpSIwx0SWqEgS5JjTDQJakRBrokNcJAl6RGGOiS1AgDXZIaYaBLUiMMdElqhIEuSY0w0CWpEQa6JDXCQJekRhjoktQIA12SGmGgS1Ij+rxT9LYkjyX5/pDxS5KcTHKwm24Yf5uSpFH6vFP008DNwGdOU/PtqnrDWDqSJK3IyDP0qvoWcGIdepEkrcK4rqG/MskDSb6e5IXDipLMJplLMjc/Pz+mXUuSYDyBfgA4v6peAnwc+PKwwqraVVUzVTUzNTU1hl1Lkp6y6kCvqser6olufi+wOcmWVXcmSVqWVQd6kucmSTd/cbfN46vdriRpeUbe5ZLkc8AlwJYkR4EPApsBquoW4ArgXUlOAb8GrqyqWrOOJUlLGhnoVXXViPGbWbitUZI0QT4pKkmNMNAlqREGuiQ1wkCXpEYY6JLUCANdkhphoEtSIwx0SWqEgS5JjTDQJakRBrokNcJAl6RGGOiS1AgDXZIaYaBLUiMMdElqhIEuSY0YGehJbkvyWJLvDxlPko8lOZLkwSQvG3+bkqRR+pyhfxq47DTjrwUu6KZZ4JOrb0uStFwjA72qvgWcOE3JTuAzteA7wDlJzhtXg5KkfsZxDX0r8Mii5aPduv8nyWySuSRz8/PzY9i1JOkp6/pH0araVVUzVTUzNTW1nruWpOaNI9CPAdsXLW/r1kmS1tE4An0P8LbubpdXACer6tExbFeStAxnjSpI8jngEmBLkqPAB4HNAFV1C7AXeB1wBPgV8I61alaSNNzIQK+qq0aMF/DusXUkSVoRnxSVpEYY6JLUCANdkhphoEtSIwx0SWqEgS5JjTDQJakRBrokNcJAl6RGGOiS1AgDXZIaYaBLUiMMdElqhIEuSY0w0CWpEQa6JDXCQJekRvQK9CSXJXk4yZEk1y0xfnWS+SQHu+md429VknQ6fd4pugn4BPAa4Chwf5I9VfXQQOmdVXXtGvQoSeqhzxn6xcCRqvphVf0WuAPYubZtSZKWq0+gbwUeWbR8tFs36M1JHkxyV5LtS20oyWySuSRz8/PzK2hXkjTMuP4o+hVguqpeDNwN7F6qqKp2VdVMVc1MTU2NadeSJOgX6MeAxWfc27p1v1dVx6vqyW7xU8BF42lPktRXn0C/H7ggyfOSPA24EtizuCDJeYsWLwcOj69FSVIfI+9yqapTSa4FvgFsAm6rqkNJbgTmqmoP8J4klwOngBPA1WvYsyRpCSMDHaCq9gJ7B9bdsGj+euD68bYmSVoOnxSVpEYY6JLUCANdkhphoEtSIwx0SWqEgS5JjTDQJakRBrokNcJAl6RGGOiS1AgDXZIaYaBLUiMMdElqhIEuSY0w0CWpEQa6JDXCQJekRvQK9CSXJXk4yZEk1y0xfnaSO7vxfUmmx96pJOm0RgZ6kk3AJ4DXAi8ArkrygoGya4CfV9XzgZuAj4y7UUnS6fU5Q78YOFJVP6yq3wJ3ADsHanYCu7v5u4BXJ8n42pQkjdLnJdFbgUcWLR8FXj6spqpOJTkJnAv8bHFRkllgtlt8IsnDK2n6DLCFgf+29ZQ2fv/xGK6Ox291NvLxO3/YQJ9AH5uq2gXsWs99roUkc1U1M+k+NjKP4ep4/Fan1ePX55LLMWD7ouVt3bola5KcBTwbOD6OBiVJ/fQJ9PuBC5I8L8nTgCuBPQM1e4C3d/NXAPdUVY2vTUnSKCMvuXTXxK8FvgFsAm6rqkNJbgTmqmoPcCtwe5IjwAkWQr9lG/6y0RnAY7g6Hr/VafL4xRNpSWqDT4pKUiMMdElqhIG+TEnemKSS/MWke9lIkpyb5GA3/TTJsUXLT5t0fxtBkuck+bckP0yyP8l/JHnTpPvaSJI8N8kdSf67O4Z7k/z5pPsaFwN9+a4C7ut+qqeqOl5VF1bVhcAtwE1PLXdPIOs0uievvwx8q6r+tKouYuHmg20TbWwD6Y7hl4BvVtWfdcfweuA5k+1sfAz0ZUjyTOCvWPjumtbv5NGZ5VLgt1V1y1MrquonVfXxCfa00fwt8D8Dx/CBqvr2BHsaKwN9eXYC/15V/wUcT3LRpBvSH4wXAgcm3cQG9yJg/6SbWEsG+vJcxcKXk9H99LKLJiLJJ5I8kOT+SfeiM8e6fpfLRpbkT1j4tfcvkxQLD1lVkn/wqVitg0PAm59aqKp3J9kCzE2upQ3nEAtPsjfLM/T+rgBur6rzq2q6qrYDPwL+esJ96Q/DPcDTk7xr0bo/nlQzG9Q9wNndt74CkOTFSZr5f9hA7+8qFv5CvtgX8LKL1kH3W+Abgb9J8qMk32XhHQT/NNHGNpDuGL4J+LvutsVDwD8DP51sZ+Pjo/+S1AjP0CWpEQa6JDXCQJekRhjoktQIA12SGmGgS1IjDHRJasT/AhUv224/wn6RAAAAAElFTkSuQmCC\n"
     },
     "metadata": {
      "needs_background": "light"
     }
    }
   ],
   "source": [
    "dna_freq = Counter(dna)\n",
    "\n",
    "plt.bar(dna_freq.keys(), dna_freq.values())"
   ]
  },
  {
   "source": [
    "# 2) Protein Synthesis"
   ],
   "cell_type": "markdown",
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Seq import Seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "Seq('MIS*')"
      ]
     },
     "metadata": {},
     "execution_count": 60
    }
   ],
   "source": [
    "dna_seq = Seq(\"ATGATCTCGTAA\") #read in 3' -> 5' direction\n",
    "len(dna_seq) #12\n",
    "\n",
    "#Complement of dna - Adenine and Thymine form double hydrogen bond pairs whereas Guanine and Cytosine form triple hydrogen bond pairs. \n",
    "dna_seq.complement() #('TACTAGAGCATT')\n",
    "\n",
    "#Reverse Compliment of dna (reading the complement strand of dna(lagging)) read in 5' -> 3' direction\n",
    "dna_seq.reverse_complement() #('TTACGAGATCAT')\n",
    "#dna_seq.complement()[::-1]\n",
    "\n",
    "\n",
    "##### PROTEIN SYNTHESIS #####\n",
    "\n",
    "#Transcription: DNA to mRNA - step simply replaces occurrences of T with U in dna_seq\n",
    "mRNA = dna_seq.transcribe()\n",
    "mRNA #AUGAUCUCGUAA\n",
    "\n",
    "#Translation: mRNA to amino acid sequence\n",
    "mRNA.translate() #('MIS*') where * is codon stop symbol and all of M I and S represent an amino acid\n",
    "\n",
    "#Can directly form amino acid from dna_seq\n",
    "dna_seq.transcribe().translate() #('MIS*') \n",
    "\n",
    "#Create custom Codon Stop Symbol\n",
    "mRNA.translate(stop_symbol=\"@\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "Seq('ATGATCTCGTAA')"
      ]
     },
     "metadata": {},
     "execution_count": 62
    }
   ],
   "source": [
    "#Back Transcription: Convert mRNA to DNA\n",
    "mRNA.back_transcribe() #('ATGATCTCGTAA')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Convert Amino Acid to 3 Letter (seq3) or 1 Letter(seq1)\n",
    "from Bio.SeqUtils import seq3,seq1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "'MIS*'"
      ]
     },
     "metadata": {},
     "execution_count": 68
    }
   ],
   "source": [
    "amino_acid = dna_seq.translate()\n",
    "seq3(amino_acid) #MetIleSerTer\n",
    "seq1(\"MetIleSerTer\") #'MIS*'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Get codon table for all of the 20 amino acids\n",
    "from Bio.Data import CodonTable\n",
    "#dir(CodonTable)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "Table 1 Standard, SGC0\n\n  |  T      |  C      |  A      |  G      |\n--+---------+---------+---------+---------+--\nT | TTT F   | TCT S   | TAT Y   | TGT C   | T\nT | TTC F   | TCC S   | TAC Y   | TGC C   | C\nT | TTA L   | TCA S   | TAA Stop| TGA Stop| A\nT | TTG L(s)| TCG S   | TAG Stop| TGG W   | G\n--+---------+---------+---------+---------+--\nC | CTT L   | CCT P   | CAT H   | CGT R   | T\nC | CTC L   | CCC P   | CAC H   | CGC R   | C\nC | CTA L   | CCA P   | CAA Q   | CGA R   | A\nC | CTG L(s)| CCG P   | CAG Q   | CGG R   | G\n--+---------+---------+---------+---------+--\nA | ATT I   | ACT T   | AAT N   | AGT S   | T\nA | ATC I   | ACC T   | AAC N   | AGC S   | C\nA | ATA I   | ACA T   | AAA K   | AGA R   | A\nA | ATG M(s)| ACG T   | AAG K   | AGG R   | G\n--+---------+---------+---------+---------+--\nG | GTT V   | GCT A   | GAT D   | GGT G   | T\nG | GTC V   | GCC A   | GAC D   | GGC G   | C\nG | GTA V   | GCA A   | GAA E   | GGA G   | A\nG | GTG V   | GCG A   | GAG E   | GGG G   | G\n--+---------+---------+---------+---------+--\n"
     ]
    }
   ],
   "source": [
    "#DNA CodonTable\n",
    "print(CodonTable.unambiguous_dna_by_name[\"Standard\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "Table 1 Standard, SGC0\n\n  |  U      |  C      |  A      |  G      |\n--+---------+---------+---------+---------+--\nU | UUU F   | UCU S   | UAU Y   | UGU C   | U\nU | UUC F   | UCC S   | UAC Y   | UGC C   | C\nU | UUA L   | UCA S   | UAA Stop| UGA Stop| A\nU | UUG L(s)| UCG S   | UAG Stop| UGG W   | G\n--+---------+---------+---------+---------+--\nC | CUU L   | CCU P   | CAU H   | CGU R   | U\nC | CUC L   | CCC P   | CAC H   | CGC R   | C\nC | CUA L   | CCA P   | CAA Q   | CGA R   | A\nC | CUG L(s)| CCG P   | CAG Q   | CGG R   | G\n--+---------+---------+---------+---------+--\nA | AUU I   | ACU T   | AAU N   | AGU S   | U\nA | AUC I   | ACC T   | AAC N   | AGC S   | C\nA | AUA I   | ACA T   | AAA K   | AGA R   | A\nA | AUG M(s)| ACG T   | AAG K   | AGG R   | G\n--+---------+---------+---------+---------+--\nG | GUU V   | GCU A   | GAU D   | GGU G   | U\nG | GUC V   | GCC A   | GAC D   | GGC G   | C\nG | GUA V   | GCA A   | GAA E   | GGA G   | A\nG | GUG V   | GCG A   | GAG E   | GGG G   | G\n--+---------+---------+---------+---------+--\n"
     ]
    }
   ],
   "source": [
    "#RNA CodonTable\n",
    "print(CodonTable.unambiguous_rna_by_name[\"Standard\"])"
   ]
  },
  {
   "source": [
    "# 3) DNA composition - AT or GC content"
   ],
   "cell_type": "markdown",
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.SeqUtils import GC\n",
    "#In Polymerase chain reaction expts, Gc content of short oligonucleotides(primers) used to print annealing temperature to template DNA.\n",
    "# Higher GC Content = higher MP = more stable = complicates PCR amplification due to difficulty in designing primer long enough to provide specificity "
   ]
  },
  {
   "source": [
    "### GC content"
   ],
   "cell_type": "markdown",
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "33.333333333333336\n"
     ]
    }
   ],
   "source": [
    "dna_seq = Seq(\"ATGATCTCGTAA\") #read in 3' -> 5' direction\n",
    "\n",
    "#GC content built in function\n",
    "GC(dna_seq) #does not matter if lowercase or uppercase\n",
    "\n",
    "#Custom functions to find GC content. Require to ensure dna sequence gets used as uppercase\n",
    "\n",
    "#1st method\n",
    "gc_content_first_method = lambda x: (x.upper().count('G') + x.upper().count('C')) / (len(dna_seq)* 1/100)\n",
    "#print(gc_content_first_method(dna_seq)) #66.7%\n",
    "\n",
    "#2nd method\n",
    "def gc_content_second_method(seq):\n",
    "    gc = [base for base in seq.upper() if base in \"GC\"]\n",
    "    return float(len(gc)) / len(dna_seq) * 100\n",
    "#print(gc_content_second_method(dna_seq)) #33.3%\n"
   ]
  },
  {
   "source": [
    "### AT Content"
   ],
   "cell_type": "markdown",
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "66.66666666666666"
      ]
     },
     "metadata": {},
     "execution_count": 110
    }
   ],
   "source": [
    "#AT content has no built in function so we have to create one\n",
    "def at_content(seq):\n",
    "    return float(seq.upper().count('A') + seq.upper().count('T')) / len(dna_seq) * 100\n",
    "\n",
    "at_content(dna_seq)"
   ]
  },
  {
   "source": [
    "### Determining Stability of DNA using its GC content where higher GC = higher Melting Point due to the stronger triple bonds\n"
   ],
   "cell_type": "markdown",
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.SeqUtils import MeltingTemp as mt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "23.32155893208184"
      ]
     },
     "metadata": {},
     "execution_count": 115
    }
   ],
   "source": [
    "#Check mp with Wallace\n",
    "mt.Tm_Wallace(dna_seq)\n",
    "\n",
    "#Check mp with GC content\n",
    "mt.Tm_GC(dna_seq) #23.32155893208184 pretty low because low GC content!"
   ]
  },
  {
   "source": [
    "### GC Skew - check when nucleotide (G,C) are over or under abundant in a particular location of dna or rna\n"
   ],
   "cell_type": "markdown",
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.SeqUtils import GC_skew,GC123,xGC_skew"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "[0.0, 0.0]"
      ]
     },
     "metadata": {},
     "execution_count": 120
    }
   ],
   "source": [
    "#GC count on 1st,2nd,3rd position\n",
    "GC123(dna_seq) #(33.333333333333336, 0.0, 25.0, 75.0)\n",
    "\n",
    "#GC skew for finding leading and lagging strands\n",
    "GC_skew(dna_seq) #[0.0] no skew with 100 windows preset\n",
    "\n",
    "GC_skew(dna_seq, window = 10) #[0.0, 0.0]\n",
    "\n",
    "#Plot GC_skew\n",
    "# xGC_skew(dna_seq) REQUIRES TKINTER!!!"
   ]
  },
  {
   "source": [
    "### Subsequences - Searching for a DNA subsequence in sequence. Returns list of [subsequence,positions]"
   ],
   "cell_type": "markdown",
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.SeqUtils import nt_search"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "['ATT', 3]"
      ]
     },
     "metadata": {},
     "execution_count": 127
    }
   ],
   "source": [
    "main_sequence = Seq(\"ACTATT\")\n",
    "subsequence = Seq(\"ATT\")\n",
    "\n",
    "nt_search(str(main_sequence),str(subsequence)) #['ATT', 3] reading a window of 3mers at a time"
   ]
  },
  {
   "source": [
    "# 4) Sequence Alignment\n",
    "\n",
    "Arranging sequences of dna, rna or amino acids or proteins to identify regions of similarity( functional, structural or evolutionary(homology - hereditary) relationships between sequences )"
   ],
   "cell_type": "markdown",
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    " from Bio import pairwise2 #2 Sequences\n",
    " from Bio.pairwise2 import format_alignment\n",
    " from Bio.Seq import Seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "seq1 = Seq('ACTCGT')\n",
    "seq2 = Seq('ATTCG')"
   ]
  },
  {
   "source": [
    "## Global Alignment - best agreement between all chars in 2 sequences\n",
    "needle\n",
    "\n",
    "\n",
    "match +1 with |\n",
    "\n",
    "mismatch . \n",
    "\n",
    "gap - "
   ],
   "cell_type": "markdown",
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "ACT-CGT\n| | || \nA-TTCG-\n  Score=4\n\n"
     ]
    }
   ],
   "source": [
    "global_alignments = pairwise2.align.globalxx(seq1,seq2) #A MATCH +1 SCORE\n",
    "#global_alignments has a weird format\n",
    "\n",
    "#Better format\n",
    "print(format_alignment(*global_alignments[0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "ACT-CGT\n| | || \nA-TTCG-\n  Score=4\n\nAC-TCGT\n|  ||| \nA-TTCG-\n  Score=4\n\nACTCGT\n|.||| \nATTCG-\n  Score=4\n\n"
     ]
    }
   ],
   "source": [
    "#View all alignments\n",
    "for a in global_alignments:\n",
    "    print(format_alignment(*a))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "4.0"
      ]
     },
     "metadata": {},
     "execution_count": 24
    }
   ],
   "source": [
    "#Get only score from alignment\n",
    "global_alignments_score = pairwise2.align.globalxx(seq1,seq2, score_only = True)\n",
    "global_alignments_score"
   ]
  },
  {
   "source": [
    "## Local Alignment - best aligning subsequence \n",
    "\n",
    "water\n",
    "\n",
    "ie - consider subsequences in each sequence and try to match them to obtain the best alignment\n",
    "\n",
    "use - \n",
    "\n",
    "2 sequences with small match region\n",
    "\n",
    "2 seq different length\n",
    "\n",
    "overlap\n",
    "\n",
    "they are subsequences of each other"
   ],
   "cell_type": "markdown",
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "1 ACT-CG\n  | | ||\n1 A-TTCG\n  Score=4\n\n1 AC-TCG\n  |  |||\n1 A-TTCG\n  Score=4\n\n1 ACTCG\n  |.|||\n1 ATTCG\n  Score=4\n\n"
     ]
    }
   ],
   "source": [
    "local_alignments = pairwise2.align.localxx(seq1,seq2)\n",
    "\n",
    "#View all alignments in better format\n",
    "for a in local_alignments:\n",
    "    print(format_alignment(*a))"
   ]
  },
  {
   "source": [
    "## Check for Similarity or %similar using Alignment\n",
    "\n",
    "### fraction of nucleotides that are the same / total no. nucleotides * 100"
   ],
   "cell_type": "markdown",
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "66.66666666666666"
      ]
     },
     "metadata": {},
     "execution_count": 26
    }
   ],
   "source": [
    "global_alignments_score/len(seq1)*100"
   ]
  },
  {
   "source": [
    "## Find out all the possible global alignments with the maximum similarity score using custom parameters\n",
    "Matching characters = 2 points\n",
    "\n",
    "Mismatching characters = -1 point\n",
    "\n",
    "open a gap = -0.5 points\n",
    "\n",
    "extending a gap = -0.1 points"
   ],
   "cell_type": "markdown",
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "ACT-CGT\n| | || \nA-TTCG-\n  Score=6.5\n\nAC-TCGT\n|  ||| \nA-TTCG-\n  Score=6.5\n\nACTCGT\n|.||| \nATTCG-\n  Score=6.5\n\n"
     ]
    }
   ],
   "source": [
    "glb_alignment = pairwise2.align.globalms(seq1,seq2,2,-1,-0.5,-0.1)\n",
    "\n",
    "for a in glb_alignment:\n",
    "    print(format_alignment(*a))"
   ]
  },
  {
   "source": [
    "# 5) Difference between Sequence Similarity and Sequence Identity"
   ],
   "cell_type": "markdown",
   "metadata": {}
  },
  {
   "source": [
    "### Similarity: \n",
    "\n",
    "resemblance between 2 sequences\n",
    "\n",
    "min. no. edits (insertion, deletion, substitution) in order to transform one sequence into an exact copy of another sequence\n",
    "\n",
    "distance (Hamming Distance etc)\n",
    "\n",
    "### Identity: \n",
    "\n",
    "no. characters match EXACTLY between 2 diff sequences\n",
    "\n",
    "gaps not counted!!!\n",
    "\n",
    "measurement is relational to shorter of the 2 sequences \n",
    "\n",
    "sequence identity is not transitive, ie: if seq A=B and B=C, A does not have to be equal to C in terms of distance"
   ],
   "cell_type": "markdown",
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Seq import Seq\n",
    "from Bio import pairwise2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    " seqA = Seq(\"AAGGCTT\")\n",
    " seqB = Seq(\"AAGGC\")\n",
    " seqC = Seq(\"AAGCAT\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    " #Compare all\n",
    " AvsB = pairwise2.align.localxx(seqA,seqB, one_alignment_only = True, score_only = True)\n",
    " BvsC = pairwise2.align.localxx(seqB,seqC, one_alignment_only = True, score_only = True)\n",
    " AvsC = pairwise2.align.localxx(seqA,seqC, one_alignment_only = True, score_only = True)\n",
    "\n",
    " #Check identity\n",
    " #print(\"AvsB\",AvsB/len(seqB)*100) #AvsB 100.0 identical because seqB perfectly fits seqA not counting gaps\n",
    " #print(\"BvsC\",BvsC/len(seqB)*100) #BvsC 100.0 identical\n",
    " #print(\"AvsC\",AvsC/len(seqC)*100)  #AvsC 83.33333333333334 identical"
   ]
  },
  {
   "source": [
    "seqA == seqB # = False, they are NOT THE SAME(SIMILAR) but they are 100% IDENTICAL "
   ],
   "cell_type": "code",
   "metadata": {},
   "execution_count": null,
   "outputs": []
  },
  {
   "source": [
    "# 6) Checking Similarity Using Distance"
   ],
   "cell_type": "markdown",
   "metadata": {}
  },
  {
   "source": [
    "## Hamming Distance \n",
    "\n",
    "Hamming distance between 2 strings of EQUAL LENGTH is the no. of positions at which the corresponding symbols are different. It returns the min. no. substitutions requried to convert 1 string into another\n",
    "\n",
    "Used in error detection and correction, quantify similarity(accounts for all types of edits and gaps) of DNA seqs, checking edit distance(method of quantifying how dissimilar 2 strings are by counting min no. operations required to transform one string into the other)\n",
    "\n",
    "## Levenshtein Distance\n",
    "\n",
    "Only accounts for indels (deletion,insertion,substitution) needed to transform 1 string to another, hence  can be used on strings that are NOT equal length"
   ],
   "cell_type": "markdown",
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "4"
      ]
     },
     "metadata": {},
     "execution_count": 55
    }
   ],
   "source": [
    "def hamming_distance(firstseq,secondseq):\n",
    "    return len([(x,y) for x,y in zip(firstseq,secondseq) if x!= y]) #zip joins both sequences together\n",
    "\n",
    "seq1 = Seq(\"ACTAT\")\n",
    "seq2 = Seq(\"ACTTA\")\n",
    "\n",
    "hamming_distance(seq1,seq2) #2 because 2 mismatches at position 3 and 4\n",
    "\n",
    "#Maximum hamming distance of any given string\n",
    "hamming_distance(seq1,seq1[::-1]) #4 ie. length of the string"
   ]
  },
  {
   "source": [
    "## Levenshtein Distance\n",
    "\n"
   ],
   "cell_type": "markdown",
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Levenshtein import distance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "Hamming Distance: 2\nLevenshtein Distance 2\n"
     ]
    }
   ],
   "source": [
    "print(\"Hamming Distance:\",hamming_distance(seq1,seq2))\n",
    "print(\"Levenshtein Distance\",distance(str(seq1),str(seq2)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "2"
      ]
     },
     "metadata": {},
     "execution_count": 63
    }
   ],
   "source": [
    "#Now with different length strings can only use Levenshtein - doesnt account for gaps, only indels\n",
    "seq3 = Seq(\"ACTGCAT\")\n",
    "distance(str(seq1),str(seq3))"
   ]
  },
  {
   "source": [
    "# 7) Checking for Similarity using DotPlot"
   ],
   "cell_type": "markdown",
   "metadata": {}
  },
  {
   "source": [
    "Dotplots are a graphical method allowing the comparison of 2 biological sequences and identify regions of close similarity (dot appears where sequences are identical -- one sequence is put on each axis and where there residuals match and the same location)\n",
    "\n",
    "Uses: \n",
    "\n",
    "direct or inverted repeats\n",
    "\n",
    "regions with low sequence complexity\n",
    "\n",
    "similar regions\n",
    "\n",
    "repeated subsequences\n",
    "\n",
    "sequence rearrangements\n",
    "\n",
    "RNA structure\n",
    "\n",
    "Gene Order "
   ],
   "cell_type": "markdown",
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Custom function for dotplot\n",
    "\n",
    "def delta(x,y):\n",
    "    return 0 if x==y else 1\n",
    " \n",
    "def M(seq4,seq5,i,j,k):\n",
    "    return sum(delta(x,y) for x,y in zip (seq4[i:i+k],seq5[j:j+k]))\n",
    " \n",
    "def makeMatrix(seq4,seq5,k):\n",
    "    n=len(seq4)\n",
    "    m=len(seq5)\n",
    "    return [[M(seq4,seq5,i,j,k) for j in range (m-k+1)] for i in range (n-k+1)]\n",
    " \n",
    "def plotMatrix(M,t,seq4,seq5,nonblank=chr(0x25A0),blank=\" \"):\n",
    "    print(\" |\"+seq5)\n",
    "    print(\"-\"*(2+len(seq5)))\n",
    "    for label,row in zip(seq4,M):\n",
    "        line=\"\".join(nonblank if s<t else blank for s in row)\n",
    "        print (label+\"|\"+line)\n",
    "        \n",
    "def dotplot(seq4,seq5,k=1,t=1):\n",
    "    M=makeMatrix(seq4,seq5,k)\n",
    "    plotMatrix(M,t,seq4,seq5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      " |AC\n----\nA|■ \nC| ■\nT|  \nT|  \nA|■ \nG|  \n"
     ]
    }
   ],
   "source": [
    "seq1 = Seq(\"ACTTAG\")\n",
    "seq2 = Seq(\"AC\")\n",
    "\n",
    "dotplot(seq1,seq2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      " |ACTTAG\n--------\nA|■   ■ \nC| ■    \nT|  ■■  \nT|  ■■  \nA|■   ■ \nG|     ■\n"
     ]
    }
   ],
   "source": [
    "#Identical = diagonal\n",
    "seq3 = seq1\n",
    "dotplot(seq1,seq3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Make more presentable\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.imshow(np.array(makeMatrix(seq1,seq2,1)))\n",
    "xt = plt.xticks(np.arange(len(list(seq1))),list(seq1)) #arange returns evenly spaced values in a given interval\n",
    "yt = plt.yticks(np.arange(len(list(seq1))),list(seq1))\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "output_type": "display_data",
     "data": {
      "text/plain": "<Figure size 432x288 with 1 Axes>",
      "image/svg+xml": "<?xml version=\"1.0\" encoding=\"utf-8\" standalone=\"no\"?>\r\n<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\r\n  \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\r\n<svg height=\"248.518125pt\" version=\"1.1\" viewBox=\"0 0 232.342656 248.518125\" width=\"232.342656pt\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\r\n <metadata>\r\n  <rdf:RDF xmlns:cc=\"http://creativecommons.org/ns#\" xmlns:dc=\"http://purl.org/dc/elements/1.1/\" xmlns:rdf=\"http://www.w3.org/1999/02/22-rdf-syntax-ns#\">\r\n   <cc:Work>\r\n    <dc:type rdf:resource=\"http://purl.org/dc/dcmitype/StillImage\"/>\r\n    <dc:date>2021-04-17T14:19:31.315962</dc:date>\r\n    <dc:format>image/svg+xml</dc:format>\r\n    <dc:creator>\r\n     <cc:Agent>\r\n      <dc:title>Matplotlib v3.4.0, https://matplotlib.org/</dc:title>\r\n     </cc:Agent>\r\n    </dc:creator>\r\n   </cc:Work>\r\n  </rdf:RDF>\r\n </metadata>\r\n <defs>\r\n  <style type=\"text/css\">*{stroke-linecap:butt;stroke-linejoin:round;}</style>\r\n </defs>\r\n <g id=\"figure_1\">\r\n  <g id=\"patch_1\">\r\n   <path d=\"M 0 248.518125 \r\nL 232.342656 248.518125 \r\nL 232.342656 0 \r\nL 0 0 \r\nz\r\n\" style=\"fill:none;\"/>\r\n  </g>\r\n  <g id=\"axes_1\">\r\n   <g id=\"patch_2\">\r\n    <path d=\"M 21.948438 224.64 \r\nL 221.268438 224.64 \r\nL 221.268438 7.2 \r\nL 21.948438 7.2 \r\nz\r\n\" style=\"fill:#ffffff;\"/>\r\n   </g>\r\n   <g clip-path=\"url(#pb902a9f90f)\">\r\n    <image height=\"218\" id=\"imagede9dddd289\" transform=\"scale(1 -1)translate(0 -218)\" width=\"73\" x=\"21.948438\" xlink:href=\"data:image/png;base64,\r\niVBORw0KGgoAAAANSUhEUgAAAEkAAADaCAYAAADqgW8xAAACBUlEQVR4nO3dwQnCUBRFwURShWsLsAtrThdWEasQXaugZ/cjzFRwObzlJ5nvt9Nj4qvD6AH/QKRApECkQKRApECkQKRApECkQKRApECkQKRApECkQKRApECkQKRApECkQKRApECkQKRApECkQKRApECkQKRApECkQKRguRzPoze8WLfr6AkfXFIgUiBSIFIgUiBSIFIgUiBSIFIgUiBSIFIgUiBSIFIgUiBSIFIgUiBSIFIgUiBSIFIgUiBSIFIgUiBSIFIgUiBSIFIgUiBSMPsY528uKRApECkQKRApECkQKRApECkQKRApECkQKRApECkQKRApECkQKRApECkQKRApECkQKRApECkQKRApECkQKRApECkQKRApECkQKRApECkQKRApECkQKRApECkQKRApECkQKRApECkQKRApECkQKRApECkQKRApECkQKRApECkQKRApECkQKRApECkQKRApWEYPeLe3H6pPk0tKRApECkQKRApECkQKRApECkQKRApECkQKRApECkQKRApECkQKRApECkQKRApECkQKRApECkQKRApECkQKRApECkQKRApECpa9PeRct+voCR9cUiBSIFIgUiBSIFIgUiBSIFIgUiBSIFIgUiBSIFIgUiBSIFIgUiBSIFIgUiBSIFIgUiBSIFIgUiBSIFIgUiBSIFIgUiBS8ARbYg5YEdjNcAAAAABJRU5ErkJggg==\" y=\"-6.64\"/>\r\n   </g>\r\n   <g id=\"matplotlib.axis_1\">\r\n    <g id=\"xtick_1\">\r\n     <g id=\"line2d_1\">\r\n      <defs>\r\n       <path d=\"M 0 0 \r\nL 0 3.5 \r\n\" id=\"mb8729371de\" style=\"stroke:#000000;stroke-width:0.8;\"/>\r\n      </defs>\r\n      <g>\r\n       <use style=\"stroke:#000000;stroke-width:0.8;\" x=\"40.068438\" xlink:href=\"#mb8729371de\" y=\"224.64\"/>\r\n      </g>\r\n     </g>\r\n     <g id=\"text_1\">\r\n      <!-- A -->\r\n      <g transform=\"translate(36.648125 239.238437)scale(0.1 -0.1)\">\r\n       <defs>\r\n        <path d=\"M 2188 4044 \r\nL 1331 1722 \r\nL 3047 1722 \r\nL 2188 4044 \r\nz\r\nM 1831 4666 \r\nL 2547 4666 \r\nL 4325 0 \r\nL 3669 0 \r\nL 3244 1197 \r\nL 1141 1197 \r\nL 716 0 \r\nL 50 0 \r\nL 1831 4666 \r\nz\r\n\" id=\"DejaVuSans-41\" transform=\"scale(0.015625)\"/>\r\n       </defs>\r\n       <use xlink:href=\"#DejaVuSans-41\"/>\r\n      </g>\r\n     </g>\r\n    </g>\r\n    <g id=\"xtick_2\">\r\n     <g id=\"line2d_2\">\r\n      <g>\r\n       <use style=\"stroke:#000000;stroke-width:0.8;\" x=\"76.308438\" xlink:href=\"#mb8729371de\" y=\"224.64\"/>\r\n      </g>\r\n     </g>\r\n     <g id=\"text_2\">\r\n      <!-- C -->\r\n      <g transform=\"translate(72.817031 239.238437)scale(0.1 -0.1)\">\r\n       <defs>\r\n        <path d=\"M 4122 4306 \r\nL 4122 3641 \r\nQ 3803 3938 3442 4084 \r\nQ 3081 4231 2675 4231 \r\nQ 1875 4231 1450 3742 \r\nQ 1025 3253 1025 2328 \r\nQ 1025 1406 1450 917 \r\nQ 1875 428 2675 428 \r\nQ 3081 428 3442 575 \r\nQ 3803 722 4122 1019 \r\nL 4122 359 \r\nQ 3791 134 3420 21 \r\nQ 3050 -91 2638 -91 \r\nQ 1578 -91 968 557 \r\nQ 359 1206 359 2328 \r\nQ 359 3453 968 4101 \r\nQ 1578 4750 2638 4750 \r\nQ 3056 4750 3426 4639 \r\nQ 3797 4528 4122 4306 \r\nz\r\n\" id=\"DejaVuSans-43\" transform=\"scale(0.015625)\"/>\r\n       </defs>\r\n       <use xlink:href=\"#DejaVuSans-43\"/>\r\n      </g>\r\n     </g>\r\n    </g>\r\n    <g id=\"xtick_3\">\r\n     <g id=\"line2d_3\">\r\n      <g>\r\n       <use style=\"stroke:#000000;stroke-width:0.8;\" x=\"112.548438\" xlink:href=\"#mb8729371de\" y=\"224.64\"/>\r\n      </g>\r\n     </g>\r\n     <g id=\"text_3\">\r\n      <!-- T -->\r\n      <g transform=\"translate(109.494531 239.238437)scale(0.1 -0.1)\">\r\n       <defs>\r\n        <path d=\"M -19 4666 \r\nL 3928 4666 \r\nL 3928 4134 \r\nL 2272 4134 \r\nL 2272 0 \r\nL 1638 0 \r\nL 1638 4134 \r\nL -19 4134 \r\nL -19 4666 \r\nz\r\n\" id=\"DejaVuSans-54\" transform=\"scale(0.015625)\"/>\r\n       </defs>\r\n       <use xlink:href=\"#DejaVuSans-54\"/>\r\n      </g>\r\n     </g>\r\n    </g>\r\n    <g id=\"xtick_4\">\r\n     <g id=\"line2d_4\">\r\n      <g>\r\n       <use style=\"stroke:#000000;stroke-width:0.8;\" x=\"148.788438\" xlink:href=\"#mb8729371de\" y=\"224.64\"/>\r\n      </g>\r\n     </g>\r\n     <g id=\"text_4\">\r\n      <!-- T -->\r\n      <g transform=\"translate(145.734531 239.238437)scale(0.1 -0.1)\">\r\n       <use xlink:href=\"#DejaVuSans-54\"/>\r\n      </g>\r\n     </g>\r\n    </g>\r\n    <g id=\"xtick_5\">\r\n     <g id=\"line2d_5\">\r\n      <g>\r\n       <use style=\"stroke:#000000;stroke-width:0.8;\" x=\"185.028438\" xlink:href=\"#mb8729371de\" y=\"224.64\"/>\r\n      </g>\r\n     </g>\r\n     <g id=\"text_5\">\r\n      <!-- A -->\r\n      <g transform=\"translate(181.608125 239.238437)scale(0.1 -0.1)\">\r\n       <use xlink:href=\"#DejaVuSans-41\"/>\r\n      </g>\r\n     </g>\r\n    </g>\r\n    <g id=\"xtick_6\">\r\n     <g id=\"line2d_6\">\r\n      <g>\r\n       <use style=\"stroke:#000000;stroke-width:0.8;\" x=\"221.268438\" xlink:href=\"#mb8729371de\" y=\"224.64\"/>\r\n      </g>\r\n     </g>\r\n     <g id=\"text_6\">\r\n      <!-- G -->\r\n      <g transform=\"translate(217.394219 239.238437)scale(0.1 -0.1)\">\r\n       <defs>\r\n        <path d=\"M 3809 666 \r\nL 3809 1919 \r\nL 2778 1919 \r\nL 2778 2438 \r\nL 4434 2438 \r\nL 4434 434 \r\nQ 4069 175 3628 42 \r\nQ 3188 -91 2688 -91 \r\nQ 1594 -91 976 548 \r\nQ 359 1188 359 2328 \r\nQ 359 3472 976 4111 \r\nQ 1594 4750 2688 4750 \r\nQ 3144 4750 3555 4637 \r\nQ 3966 4525 4313 4306 \r\nL 4313 3634 \r\nQ 3963 3931 3569 4081 \r\nQ 3175 4231 2741 4231 \r\nQ 1884 4231 1454 3753 \r\nQ 1025 3275 1025 2328 \r\nQ 1025 1384 1454 906 \r\nQ 1884 428 2741 428 \r\nQ 3075 428 3337 486 \r\nQ 3600 544 3809 666 \r\nz\r\n\" id=\"DejaVuSans-47\" transform=\"scale(0.015625)\"/>\r\n       </defs>\r\n       <use xlink:href=\"#DejaVuSans-47\"/>\r\n      </g>\r\n     </g>\r\n    </g>\r\n   </g>\r\n   <g id=\"matplotlib.axis_2\">\r\n    <g id=\"ytick_1\">\r\n     <g id=\"line2d_7\">\r\n      <defs>\r\n       <path d=\"M 0 0 \r\nL -3.5 0 \r\n\" id=\"m1b980d8951\" style=\"stroke:#000000;stroke-width:0.8;\"/>\r\n      </defs>\r\n      <g>\r\n       <use style=\"stroke:#000000;stroke-width:0.8;\" x=\"21.948438\" xlink:href=\"#m1b980d8951\" y=\"25.32\"/>\r\n      </g>\r\n     </g>\r\n     <g id=\"text_7\">\r\n      <!-- A -->\r\n      <g transform=\"translate(8.107813 29.119219)scale(0.1 -0.1)\">\r\n       <use xlink:href=\"#DejaVuSans-41\"/>\r\n      </g>\r\n     </g>\r\n    </g>\r\n    <g id=\"ytick_2\">\r\n     <g id=\"line2d_8\">\r\n      <g>\r\n       <use style=\"stroke:#000000;stroke-width:0.8;\" x=\"21.948438\" xlink:href=\"#m1b980d8951\" y=\"61.56\"/>\r\n      </g>\r\n     </g>\r\n     <g id=\"text_8\">\r\n      <!-- C -->\r\n      <g transform=\"translate(7.965625 65.359219)scale(0.1 -0.1)\">\r\n       <use xlink:href=\"#DejaVuSans-43\"/>\r\n      </g>\r\n     </g>\r\n    </g>\r\n    <g id=\"ytick_3\">\r\n     <g id=\"line2d_9\">\r\n      <g>\r\n       <use style=\"stroke:#000000;stroke-width:0.8;\" x=\"21.948438\" xlink:href=\"#m1b980d8951\" y=\"97.8\"/>\r\n      </g>\r\n     </g>\r\n     <g id=\"text_9\">\r\n      <!-- T -->\r\n      <g transform=\"translate(8.840625 101.599219)scale(0.1 -0.1)\">\r\n       <use xlink:href=\"#DejaVuSans-54\"/>\r\n      </g>\r\n     </g>\r\n    </g>\r\n    <g id=\"ytick_4\">\r\n     <g id=\"line2d_10\">\r\n      <g>\r\n       <use style=\"stroke:#000000;stroke-width:0.8;\" x=\"21.948438\" xlink:href=\"#m1b980d8951\" y=\"134.04\"/>\r\n      </g>\r\n     </g>\r\n     <g id=\"text_10\">\r\n      <!-- T -->\r\n      <g transform=\"translate(8.840625 137.839219)scale(0.1 -0.1)\">\r\n       <use xlink:href=\"#DejaVuSans-54\"/>\r\n      </g>\r\n     </g>\r\n    </g>\r\n    <g id=\"ytick_5\">\r\n     <g id=\"line2d_11\">\r\n      <g>\r\n       <use style=\"stroke:#000000;stroke-width:0.8;\" x=\"21.948438\" xlink:href=\"#m1b980d8951\" y=\"170.28\"/>\r\n      </g>\r\n     </g>\r\n     <g id=\"text_11\">\r\n      <!-- A -->\r\n      <g transform=\"translate(8.107813 174.079219)scale(0.1 -0.1)\">\r\n       <use xlink:href=\"#DejaVuSans-41\"/>\r\n      </g>\r\n     </g>\r\n    </g>\r\n    <g id=\"ytick_6\">\r\n     <g id=\"line2d_12\">\r\n      <g>\r\n       <use style=\"stroke:#000000;stroke-width:0.8;\" x=\"21.948438\" xlink:href=\"#m1b980d8951\" y=\"206.52\"/>\r\n      </g>\r\n     </g>\r\n     <g id=\"text_12\">\r\n      <!-- G -->\r\n      <g transform=\"translate(7.2 210.319219)scale(0.1 -0.1)\">\r\n       <use xlink:href=\"#DejaVuSans-47\"/>\r\n      </g>\r\n     </g>\r\n    </g>\r\n   </g>\r\n   <g id=\"patch_3\">\r\n    <path d=\"M 21.948438 224.64 \r\nL 21.948438 7.2 \r\n\" style=\"fill:none;stroke:#000000;stroke-linecap:square;stroke-linejoin:miter;stroke-width:0.8;\"/>\r\n   </g>\r\n   <g id=\"patch_4\">\r\n    <path d=\"M 221.268438 224.64 \r\nL 221.268438 7.2 \r\n\" style=\"fill:none;stroke:#000000;stroke-linecap:square;stroke-linejoin:miter;stroke-width:0.8;\"/>\r\n   </g>\r\n   <g id=\"patch_5\">\r\n    <path d=\"M 21.948438 224.64 \r\nL 221.268438 224.64 \r\n\" style=\"fill:none;stroke:#000000;stroke-linecap:square;stroke-linejoin:miter;stroke-width:0.8;\"/>\r\n   </g>\r\n   <g id=\"patch_6\">\r\n    <path d=\"M 21.948438 7.2 \r\nL 221.268438 7.2 \r\n\" style=\"fill:none;stroke:#000000;stroke-linecap:square;stroke-linejoin:miter;stroke-width:0.8;\"/>\r\n   </g>\r\n  </g>\r\n </g>\r\n <defs>\r\n  <clipPath id=\"pb902a9f90f\">\r\n   <rect height=\"217.44\" width=\"199.32\" x=\"21.948438\" y=\"7.2\"/>\r\n  </clipPath>\r\n </defs>\r\n</svg>\r\n",
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAOgAAAD4CAYAAADmf6rjAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjQuMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8+yak3AAAACXBIWXMAAAsTAAALEwEAmpwYAAAIrElEQVR4nO3dQYgdhR3H8d/PrUbU9mAMBqlEekiFagzNYi+GEhW81UhDyZ6FheJZemxuXjxZCpKbCGV7EHsSQzGHVlpoX0os5BKwwUOtsO5NBA3pr4e81W2yyc4mb3w/ed8PhOxMHpP/Yb6ZyXvzZpxEADrdMe8BANwYgQLFCBQoRqBAMQIFin1nrA3f5T25W/eOtfldO3jo83mPcL07H5v3BJiDc+fOfZpk35DXjhbo3bpXP/EzY21+186cOT/vEa5zx/7JvEfAHNj+aOhrOcUFihEoUIxAgWIEChQjUKAYgQLFCBQoRqBAMQIFihEoUIxAgWIEChQjUKDY4EBtH7cd24+OORCAr+3mCLoi6f3p7wC+AYMCtX2fpKckvSjp5KgTAfjK0CPo85LeTXJR0obtI9u9yPaq7YntyWV9MbMhgUU1NNAVSWvTn9d0g9PcJKeTLCdZvlN7ZjEfsNB2vOWJ7fslPS3pcduRtCQptl8Ot6UHRjXkCHpC0ptJDiR5JMnDki5JOjruaACGBLoi6e1r1r0l3s0FRrfjKW6SY9use22ccQBsxZVEQDECBYoRKFCMQIFiBAoUI1CgGIECxQgUKEagQDECBYoRKFCMQIFiO14sf6sOHvq86rHzzz10eN4jXOeP/533BGjHERQoRqBAMQIFihEoUIxAgWIEChQjUKAYgQLFCBQoRqBAMQIFihEoUIxAgWIEChQjUKDY4EBt77e9ZvtD2+dsv2P74JjDAYtu0Be2bVtXH0H4RpKT03VPSHpQ0sXxxgMW29A7KhyTdDnJ65srknwwzkgANg09xX1M0rmdXmR71fbE9mR948rtTQZgtm8SJTmdZDnJ8r69S7PcNLCQhgZ6QdKRMQcBcL2hgZ6VtMf26uYK24dsHx1nLADSwECTRNILkp6dfsxyQdIrkj4Zczhg0Q2+L26SjyX9YsRZAFyDK4mAYgQKFCNQoBiBAsUIFChGoEAxAgWKEShQjECBYgQKFCNQoBiBAsUGXyz/bXfm4/PzHgHYNY6gQDECBYoRKFCMQIFiBAoUI1CgGIECxQgUKEagQDECBYoRKFCMQIFiBAoUI1Cg2KCvm9neK+m96eJ+SVckrU+Xn0zy5QizAQtvUKBJNiQdliTbpyR9luTV8cYCIHGKC1SbaaC2V21PbE/WN67MctPAQpppoElOJ1lOsrxv79IsNw0sJE5xgWIEChQjUKDYrm+7meTUCHMA2AZHUKAYgQLFCBQoRqBAMQIFihEoUIxAgWIEChQjUKAYgQLFCBQoRqBAMQIFihEoUIxAgWIEChQjUKAYgQLFCBQoRqBAMQIFihEoUIxAgWIEChQjUKAYgQLFCBQoNujZLLb3Snpvurhf0hVJ69PlJ5N8OcJswMIbFGiSDUmHJcn2KUmfJXl1vLEASJziAtVmGqjtVdsT25P1jSuz3DSwkGYaaJLTSZaTLO/buzTLTQMLiVNcoBiBAsUIFCg26GOWrZKcGmEOANvgCAoUI1CgGIECxQgUKEagQDECBYoRKFCMQIFiBAoUI1CgGIECxQgUKLbri+WHuvjPe/TcQ4fH2vyunfn4/LxHAHaNIyhQjECBYgQKFCNQoBiBAsUIFChGoEAxAgWKEShQjECBYgQKFCNQoBiBAsUIFCg2OFDbx23H9qNjDgTga7s5gq5Ien/6O4BvwKBAbd8n6SlJL0o6OepEAL4y9Aj6vKR3k1yUtGH7yHYvsr1qe2J7cllfzGxIYFENDXRF0tr05zXd4DQ3yekky0mW79SeWcwHLLQd70lk+35JT0t63HYkLUmK7ZeTZOwBgUU25Ah6QtKbSQ4keSTJw5IuSTo67mgAhgS6Iunta9a9Jd7NBUa34ylukmPbrHttnHEAbMWVREAxAgWKEShQjECBYgQKFCNQoBiBAsUIFChGoEAxAgWKEShQjECBYgQKFNvx2yy36uChz3XmzPmxNg8sBI6gQDECBYoRKFCMQIFiBAoUI1CgGIECxQgUKEagQDECBYoRKFCMQIFiBAoUI1Cg2KBAbT9o+3e2/2X7nO2/2n5h7OGARbdjoLYt6Q+S/pTkB0mOSDop6fsjzwYsvCFf2H5a0pdJXt9ckeQjSb8ZbSoAkoad4v5I0j+GbMz2qu2J7cn6xpXbmwzA7t8ksv1b2x/Y/vu1f5bkdJLlJMv79i7NZkJggQ0J9IKkH28uJHlJ0jOS9o01FICrhgR6VtLdtn+5Zd09I80DYIsdA00SSccl/dT2Jdt/k/SGpF+NPBuw8AbddjPJf3T1oxUA3yCuJAKKEShQjECBYgQKFCNQoBiBAsUIFChGoEAxAgWKEShQjECBYgQKFPPVL6uMsGF7XdJHM9jUA5I+ncF2ZqVtHqlvJua5uR8m+e6QFw76NsutSDKTL3TbniRZnsW2ZqFtHqlvJua5OduToa/lFBcoRqBAsW9DoKfnPcA12uaR+mZinpsbPM9obxIBuH3fhiMosLAIFChWHajt47Zj+9GCWfbbXrP94fQBUu/YPjinWfbaPj/99Yntf29ZvouZ/m+upn1o1w8hq/4/qO3fS3pI0tkkv57jHJb0F0lvbD6jxvYTkr6X5M/zmms6xylJnyV5dZ5zbNU0U/k+dEDSz5Lc8DlHtUdQ2/dJekrSi5r/LT+PSbp8zQOkPph3nLi5sn1o24eQ3SxOqThQSc9LejfJRUkbto/McZbHJJ2b49+PW9O0Dw1+CNlWzYGuSFqb/rw2XQZ2o3YfutlDyLYa7Vrc22H7fl09JXjcdiQtSYrtlzOf/zRfkHRiDn8vblHpPvTzzYUkL9l+QNJNr8ttPYKekPRmkgNJHknysKRLko7OaZ6zkvbYXt1cYfuQ7XnNg5017kO7fghZa6Arkt6+Zt1bmtMpyvRf3BckPTv9mOWCpFckfTKPeTBI4z50XLt8CFn1xyzAoms9ggIQgQLVCBQoRqBAMQIFihEoUIxAgWL/A03rGDWL8ZnuAAAAAElFTkSuQmCC\n"
     },
     "metadata": {
      "needs_background": "light"
     }
    }
   ],
   "source": [
    "#Final product \n",
    "def dotplotx(seq1,seq2):\n",
    "    plt.imshow(np.array(makeMatrix(seq1,seq2,1)))\n",
    "    xt = plt.xticks(np.arange(len(list(seq1))),list(seq1)) #arange returns evenly spaced values in a given interval\n",
    "    yt = plt.yticks(np.arange(len(list(seq1))),list(seq1))\n",
    "    plt.show()\n",
    "\n",
    "dotplotx(seq1,seq2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "output_type": "display_data",
     "data": {
      "text/plain": "<Figure size 432x288 with 1 Axes>",
      "image/svg+xml": "<?xml version=\"1.0\" encoding=\"utf-8\" standalone=\"no\"?>\r\n<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\r\n  \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\r\n<svg height=\"248.518125pt\" version=\"1.1\" viewBox=\"0 0 246.588437 248.518125\" width=\"246.588437pt\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\r\n <metadata>\r\n  <rdf:RDF xmlns:cc=\"http://creativecommons.org/ns#\" xmlns:dc=\"http://purl.org/dc/elements/1.1/\" xmlns:rdf=\"http://www.w3.org/1999/02/22-rdf-syntax-ns#\">\r\n   <cc:Work>\r\n    <dc:type rdf:resource=\"http://purl.org/dc/dcmitype/StillImage\"/>\r\n    <dc:date>2021-04-17T14:22:36.773661</dc:date>\r\n    <dc:format>image/svg+xml</dc:format>\r\n    <dc:creator>\r\n     <cc:Agent>\r\n      <dc:title>Matplotlib v3.4.0, https://matplotlib.org/</dc:title>\r\n     </cc:Agent>\r\n    </dc:creator>\r\n   </cc:Work>\r\n  </rdf:RDF>\r\n </metadata>\r\n <defs>\r\n  <style type=\"text/css\">*{stroke-linecap:butt;stroke-linejoin:round;}</style>\r\n </defs>\r\n <g id=\"figure_1\">\r\n  <g id=\"patch_1\">\r\n   <path d=\"M 0 248.518125 \r\nL 246.588437 248.518125 \r\nL 246.588437 0 \r\nL 0 0 \r\nz\r\n\" style=\"fill:none;\"/>\r\n  </g>\r\n  <g id=\"axes_1\">\r\n   <g id=\"patch_2\">\r\n    <path d=\"M 21.948438 224.64 \r\nL 239.388438 224.64 \r\nL 239.388438 7.2 \r\nL 21.948438 7.2 \r\nz\r\n\" style=\"fill:#ffffff;\"/>\r\n   </g>\r\n   <g clip-path=\"url(#p0de5f4141f)\">\r\n    <image height=\"218\" id=\"image89aabaeaa3\" transform=\"scale(1 -1)translate(0 -218)\" width=\"218\" x=\"21.948438\" xlink:href=\"data:image/png;base64,\r\niVBORw0KGgoAAAANSUhEUgAAANoAAADaCAYAAADAHVzbAAADL0lEQVR4nO3d0WkbURBA0WxIFflWAS5CkAbUbLpIFXYVxu5Aa/BwM0jnFLDsQ7q8r9k5rsft48eAv6//Jh4z5s/vl7FnbTsb5yZ//wk///cLwDMQGgSEBgGhQUBoEBAaBIQGAaFBQGgQEBoEhAYBoUFAaBAQGgSEBgGhQUBoEBAaBIQGAaFBQGgQEBoEhAYBoUFAaBAQGgSEBgGhQUBoEBAaBIQGAaFBQGgQEBoEhAaB4/3tMrLDetvO4Mm901Nnm3qnbe8z5ZH3jrvRICA0CAgNAkKDgNAgIDQICA0CQoOA0CAgNAgIDQJCg4DQICA0CAgNAkKDgNAgIDQIjH3KgHM+QXDftnNNcqNBQGgQEBoEhAYBoUFAaBAQGgSEBgGhQUBoEBAaBIQGAaFBQGgQEBoEhAYBoUHAhPUTMxndcaNBQGgQEBoEhAYBoUFAaBAQGgSEBgGhQUBoEBAaBIQGAaFBQGgQEBoEhAYBoUHg19SDpqZ1p0xO/W6bRN72PlMm/0PbzuZGg4DQICA0CAgNAkKDgNAgIDQICA0CQoOA0CAgNAgIDQJCg4DQICA0CAgNAkKDgB3WT+xRJ7U3cqNBQGgQEBoEhAYBoUFAaBAQGgSEBgGhQUBoEBAaBIQGAaFBQGgQEBoEhAYBoUHADusv2DaJvO19pthhDXyL0CAgNAgIDQJCg4DQICA0CAgNAkKDgNAgIDQICA0CQoOA0CAgNAgIDQJCg8BxPW4jO6y3TbQ+8rQu57ZN/LvRICA0CAgNAkKDgNAgIDQICA0CQoOA0CAgNAgIDQJCg4DQICA0CAgNAkKDgNAgIDQIHO9vl5FPGXDOkvf7tp1rkhsNAkKDgNAgIDQICA0CQoOA0CAgNAgIDQJCg4DQICA0CAgNAkKDgNAgIDQICA0CYxPW25ZzT07rbpsg3vY+Uyb/Q9vO5kaDgNAgIDQICA0CQoOA0CAgNAgIDQJCg4DQICA0CAgNAkKDgNAgIDQICA0CQoPAcT1uIxPW2yZaH3lal3PbJv7daBAQGgSEBgGhQUBoEBAaBIQGAaFBQGgQEBoEhAYBoUFAaBAQGgSEBgGhQUBoEPgEcdNsoqspMZEAAAAASUVORK5CYII=\" y=\"-6.64\"/>\r\n   </g>\r\n   <g id=\"matplotlib.axis_1\">\r\n    <g id=\"xtick_1\">\r\n     <g id=\"line2d_1\">\r\n      <defs>\r\n       <path d=\"M 0 0 \r\nL 0 3.5 \r\n\" id=\"mb1ef57c73a\" style=\"stroke:#000000;stroke-width:0.8;\"/>\r\n      </defs>\r\n      <g>\r\n       <use style=\"stroke:#000000;stroke-width:0.8;\" x=\"31.008438\" xlink:href=\"#mb1ef57c73a\" y=\"224.64\"/>\r\n      </g>\r\n     </g>\r\n     <g id=\"text_1\">\r\n      <!-- A -->\r\n      <g transform=\"translate(27.588125 239.238437)scale(0.1 -0.1)\">\r\n       <defs>\r\n        <path d=\"M 2188 4044 \r\nL 1331 1722 \r\nL 3047 1722 \r\nL 2188 4044 \r\nz\r\nM 1831 4666 \r\nL 2547 4666 \r\nL 4325 0 \r\nL 3669 0 \r\nL 3244 1197 \r\nL 1141 1197 \r\nL 716 0 \r\nL 50 0 \r\nL 1831 4666 \r\nz\r\n\" id=\"DejaVuSans-41\" transform=\"scale(0.015625)\"/>\r\n       </defs>\r\n       <use xlink:href=\"#DejaVuSans-41\"/>\r\n      </g>\r\n     </g>\r\n    </g>\r\n    <g id=\"xtick_2\">\r\n     <g id=\"line2d_2\">\r\n      <g>\r\n       <use style=\"stroke:#000000;stroke-width:0.8;\" x=\"49.128438\" xlink:href=\"#mb1ef57c73a\" y=\"224.64\"/>\r\n      </g>\r\n     </g>\r\n     <g id=\"text_2\">\r\n      <!-- T -->\r\n      <g transform=\"translate(46.074531 239.238437)scale(0.1 -0.1)\">\r\n       <defs>\r\n        <path d=\"M -19 4666 \r\nL 3928 4666 \r\nL 3928 4134 \r\nL 2272 4134 \r\nL 2272 0 \r\nL 1638 0 \r\nL 1638 4134 \r\nL -19 4134 \r\nL -19 4666 \r\nz\r\n\" id=\"DejaVuSans-54\" transform=\"scale(0.015625)\"/>\r\n       </defs>\r\n       <use xlink:href=\"#DejaVuSans-54\"/>\r\n      </g>\r\n     </g>\r\n    </g>\r\n    <g id=\"xtick_3\">\r\n     <g id=\"line2d_3\">\r\n      <g>\r\n       <use style=\"stroke:#000000;stroke-width:0.8;\" x=\"67.248438\" xlink:href=\"#mb1ef57c73a\" y=\"224.64\"/>\r\n      </g>\r\n     </g>\r\n     <g id=\"text_3\">\r\n      <!-- G -->\r\n      <g transform=\"translate(63.374219 239.238437)scale(0.1 -0.1)\">\r\n       <defs>\r\n        <path d=\"M 3809 666 \r\nL 3809 1919 \r\nL 2778 1919 \r\nL 2778 2438 \r\nL 4434 2438 \r\nL 4434 434 \r\nQ 4069 175 3628 42 \r\nQ 3188 -91 2688 -91 \r\nQ 1594 -91 976 548 \r\nQ 359 1188 359 2328 \r\nQ 359 3472 976 4111 \r\nQ 1594 4750 2688 4750 \r\nQ 3144 4750 3555 4637 \r\nQ 3966 4525 4313 4306 \r\nL 4313 3634 \r\nQ 3963 3931 3569 4081 \r\nQ 3175 4231 2741 4231 \r\nQ 1884 4231 1454 3753 \r\nQ 1025 3275 1025 2328 \r\nQ 1025 1384 1454 906 \r\nQ 1884 428 2741 428 \r\nQ 3075 428 3337 486 \r\nQ 3600 544 3809 666 \r\nz\r\n\" id=\"DejaVuSans-47\" transform=\"scale(0.015625)\"/>\r\n       </defs>\r\n       <use xlink:href=\"#DejaVuSans-47\"/>\r\n      </g>\r\n     </g>\r\n    </g>\r\n    <g id=\"xtick_4\">\r\n     <g id=\"line2d_4\">\r\n      <g>\r\n       <use style=\"stroke:#000000;stroke-width:0.8;\" x=\"85.368437\" xlink:href=\"#mb1ef57c73a\" y=\"224.64\"/>\r\n      </g>\r\n     </g>\r\n     <g id=\"text_4\">\r\n      <!-- A -->\r\n      <g transform=\"translate(81.948125 239.238437)scale(0.1 -0.1)\">\r\n       <use xlink:href=\"#DejaVuSans-41\"/>\r\n      </g>\r\n     </g>\r\n    </g>\r\n    <g id=\"xtick_5\">\r\n     <g id=\"line2d_5\">\r\n      <g>\r\n       <use style=\"stroke:#000000;stroke-width:0.8;\" x=\"103.488438\" xlink:href=\"#mb1ef57c73a\" y=\"224.64\"/>\r\n      </g>\r\n     </g>\r\n     <g id=\"text_5\">\r\n      <!-- T -->\r\n      <g transform=\"translate(100.434531 239.238437)scale(0.1 -0.1)\">\r\n       <use xlink:href=\"#DejaVuSans-54\"/>\r\n      </g>\r\n     </g>\r\n    </g>\r\n    <g id=\"xtick_6\">\r\n     <g id=\"line2d_6\">\r\n      <g>\r\n       <use style=\"stroke:#000000;stroke-width:0.8;\" x=\"121.608438\" xlink:href=\"#mb1ef57c73a\" y=\"224.64\"/>\r\n      </g>\r\n     </g>\r\n     <g id=\"text_6\">\r\n      <!-- C -->\r\n      <g transform=\"translate(118.117031 239.238437)scale(0.1 -0.1)\">\r\n       <defs>\r\n        <path d=\"M 4122 4306 \r\nL 4122 3641 \r\nQ 3803 3938 3442 4084 \r\nQ 3081 4231 2675 4231 \r\nQ 1875 4231 1450 3742 \r\nQ 1025 3253 1025 2328 \r\nQ 1025 1406 1450 917 \r\nQ 1875 428 2675 428 \r\nQ 3081 428 3442 575 \r\nQ 3803 722 4122 1019 \r\nL 4122 359 \r\nQ 3791 134 3420 21 \r\nQ 3050 -91 2638 -91 \r\nQ 1578 -91 968 557 \r\nQ 359 1206 359 2328 \r\nQ 359 3453 968 4101 \r\nQ 1578 4750 2638 4750 \r\nQ 3056 4750 3426 4639 \r\nQ 3797 4528 4122 4306 \r\nz\r\n\" id=\"DejaVuSans-43\" transform=\"scale(0.015625)\"/>\r\n       </defs>\r\n       <use xlink:href=\"#DejaVuSans-43\"/>\r\n      </g>\r\n     </g>\r\n    </g>\r\n    <g id=\"xtick_7\">\r\n     <g id=\"line2d_7\">\r\n      <g>\r\n       <use style=\"stroke:#000000;stroke-width:0.8;\" x=\"139.728437\" xlink:href=\"#mb1ef57c73a\" y=\"224.64\"/>\r\n      </g>\r\n     </g>\r\n     <g id=\"text_7\">\r\n      <!-- T -->\r\n      <g transform=\"translate(136.674531 239.238437)scale(0.1 -0.1)\">\r\n       <use xlink:href=\"#DejaVuSans-54\"/>\r\n      </g>\r\n     </g>\r\n    </g>\r\n    <g id=\"xtick_8\">\r\n     <g id=\"line2d_8\">\r\n      <g>\r\n       <use style=\"stroke:#000000;stroke-width:0.8;\" x=\"157.848437\" xlink:href=\"#mb1ef57c73a\" y=\"224.64\"/>\r\n      </g>\r\n     </g>\r\n     <g id=\"text_8\">\r\n      <!-- C -->\r\n      <g transform=\"translate(154.357031 239.238437)scale(0.1 -0.1)\">\r\n       <use xlink:href=\"#DejaVuSans-43\"/>\r\n      </g>\r\n     </g>\r\n    </g>\r\n    <g id=\"xtick_9\">\r\n     <g id=\"line2d_9\">\r\n      <g>\r\n       <use style=\"stroke:#000000;stroke-width:0.8;\" x=\"175.968437\" xlink:href=\"#mb1ef57c73a\" y=\"224.64\"/>\r\n      </g>\r\n     </g>\r\n     <g id=\"text_9\">\r\n      <!-- G -->\r\n      <g transform=\"translate(172.094219 239.238437)scale(0.1 -0.1)\">\r\n       <use xlink:href=\"#DejaVuSans-47\"/>\r\n      </g>\r\n     </g>\r\n    </g>\r\n    <g id=\"xtick_10\">\r\n     <g id=\"line2d_10\">\r\n      <g>\r\n       <use style=\"stroke:#000000;stroke-width:0.8;\" x=\"194.088437\" xlink:href=\"#mb1ef57c73a\" y=\"224.64\"/>\r\n      </g>\r\n     </g>\r\n     <g id=\"text_10\">\r\n      <!-- T -->\r\n      <g transform=\"translate(191.034531 239.238437)scale(0.1 -0.1)\">\r\n       <use xlink:href=\"#DejaVuSans-54\"/>\r\n      </g>\r\n     </g>\r\n    </g>\r\n    <g id=\"xtick_11\">\r\n     <g id=\"line2d_11\">\r\n      <g>\r\n       <use style=\"stroke:#000000;stroke-width:0.8;\" x=\"212.208438\" xlink:href=\"#mb1ef57c73a\" y=\"224.64\"/>\r\n      </g>\r\n     </g>\r\n     <g id=\"text_11\">\r\n      <!-- A -->\r\n      <g transform=\"translate(208.788125 239.238437)scale(0.1 -0.1)\">\r\n       <use xlink:href=\"#DejaVuSans-41\"/>\r\n      </g>\r\n     </g>\r\n    </g>\r\n    <g id=\"xtick_12\">\r\n     <g id=\"line2d_12\">\r\n      <g>\r\n       <use style=\"stroke:#000000;stroke-width:0.8;\" x=\"230.328437\" xlink:href=\"#mb1ef57c73a\" y=\"224.64\"/>\r\n      </g>\r\n     </g>\r\n     <g id=\"text_12\">\r\n      <!-- A -->\r\n      <g transform=\"translate(226.908125 239.238437)scale(0.1 -0.1)\">\r\n       <use xlink:href=\"#DejaVuSans-41\"/>\r\n      </g>\r\n     </g>\r\n    </g>\r\n   </g>\r\n   <g id=\"matplotlib.axis_2\">\r\n    <g id=\"ytick_1\">\r\n     <g id=\"line2d_13\">\r\n      <defs>\r\n       <path d=\"M 0 0 \r\nL -3.5 0 \r\n\" id=\"m083fbb6f0c\" style=\"stroke:#000000;stroke-width:0.8;\"/>\r\n      </defs>\r\n      <g>\r\n       <use style=\"stroke:#000000;stroke-width:0.8;\" x=\"21.948438\" xlink:href=\"#m083fbb6f0c\" y=\"16.26\"/>\r\n      </g>\r\n     </g>\r\n     <g id=\"text_13\">\r\n      <!-- A -->\r\n      <g transform=\"translate(8.107813 20.059219)scale(0.1 -0.1)\">\r\n       <use xlink:href=\"#DejaVuSans-41\"/>\r\n      </g>\r\n     </g>\r\n    </g>\r\n    <g id=\"ytick_2\">\r\n     <g id=\"line2d_14\">\r\n      <g>\r\n       <use style=\"stroke:#000000;stroke-width:0.8;\" x=\"21.948438\" xlink:href=\"#m083fbb6f0c\" y=\"34.38\"/>\r\n      </g>\r\n     </g>\r\n     <g id=\"text_14\">\r\n      <!-- T -->\r\n      <g transform=\"translate(8.840625 38.179219)scale(0.1 -0.1)\">\r\n       <use xlink:href=\"#DejaVuSans-54\"/>\r\n      </g>\r\n     </g>\r\n    </g>\r\n    <g id=\"ytick_3\">\r\n     <g id=\"line2d_15\">\r\n      <g>\r\n       <use style=\"stroke:#000000;stroke-width:0.8;\" x=\"21.948438\" xlink:href=\"#m083fbb6f0c\" y=\"52.5\"/>\r\n      </g>\r\n     </g>\r\n     <g id=\"text_15\">\r\n      <!-- G -->\r\n      <g transform=\"translate(7.2 56.299219)scale(0.1 -0.1)\">\r\n       <use xlink:href=\"#DejaVuSans-47\"/>\r\n      </g>\r\n     </g>\r\n    </g>\r\n    <g id=\"ytick_4\">\r\n     <g id=\"line2d_16\">\r\n      <g>\r\n       <use style=\"stroke:#000000;stroke-width:0.8;\" x=\"21.948438\" xlink:href=\"#m083fbb6f0c\" y=\"70.62\"/>\r\n      </g>\r\n     </g>\r\n     <g id=\"text_16\">\r\n      <!-- A -->\r\n      <g transform=\"translate(8.107813 74.419219)scale(0.1 -0.1)\">\r\n       <use xlink:href=\"#DejaVuSans-41\"/>\r\n      </g>\r\n     </g>\r\n    </g>\r\n    <g id=\"ytick_5\">\r\n     <g id=\"line2d_17\">\r\n      <g>\r\n       <use style=\"stroke:#000000;stroke-width:0.8;\" x=\"21.948438\" xlink:href=\"#m083fbb6f0c\" y=\"88.74\"/>\r\n      </g>\r\n     </g>\r\n     <g id=\"text_17\">\r\n      <!-- T -->\r\n      <g transform=\"translate(8.840625 92.539219)scale(0.1 -0.1)\">\r\n       <use xlink:href=\"#DejaVuSans-54\"/>\r\n      </g>\r\n     </g>\r\n    </g>\r\n    <g id=\"ytick_6\">\r\n     <g id=\"line2d_18\">\r\n      <g>\r\n       <use style=\"stroke:#000000;stroke-width:0.8;\" x=\"21.948438\" xlink:href=\"#m083fbb6f0c\" y=\"106.86\"/>\r\n      </g>\r\n     </g>\r\n     <g id=\"text_18\">\r\n      <!-- C -->\r\n      <g transform=\"translate(7.965625 110.659219)scale(0.1 -0.1)\">\r\n       <use xlink:href=\"#DejaVuSans-43\"/>\r\n      </g>\r\n     </g>\r\n    </g>\r\n    <g id=\"ytick_7\">\r\n     <g id=\"line2d_19\">\r\n      <g>\r\n       <use style=\"stroke:#000000;stroke-width:0.8;\" x=\"21.948438\" xlink:href=\"#m083fbb6f0c\" y=\"124.98\"/>\r\n      </g>\r\n     </g>\r\n     <g id=\"text_19\">\r\n      <!-- T -->\r\n      <g transform=\"translate(8.840625 128.779219)scale(0.1 -0.1)\">\r\n       <use xlink:href=\"#DejaVuSans-54\"/>\r\n      </g>\r\n     </g>\r\n    </g>\r\n    <g id=\"ytick_8\">\r\n     <g id=\"line2d_20\">\r\n      <g>\r\n       <use style=\"stroke:#000000;stroke-width:0.8;\" x=\"21.948438\" xlink:href=\"#m083fbb6f0c\" y=\"143.1\"/>\r\n      </g>\r\n     </g>\r\n     <g id=\"text_20\">\r\n      <!-- C -->\r\n      <g transform=\"translate(7.965625 146.899219)scale(0.1 -0.1)\">\r\n       <use xlink:href=\"#DejaVuSans-43\"/>\r\n      </g>\r\n     </g>\r\n    </g>\r\n    <g id=\"ytick_9\">\r\n     <g id=\"line2d_21\">\r\n      <g>\r\n       <use style=\"stroke:#000000;stroke-width:0.8;\" x=\"21.948438\" xlink:href=\"#m083fbb6f0c\" y=\"161.22\"/>\r\n      </g>\r\n     </g>\r\n     <g id=\"text_21\">\r\n      <!-- G -->\r\n      <g transform=\"translate(7.2 165.019219)scale(0.1 -0.1)\">\r\n       <use xlink:href=\"#DejaVuSans-47\"/>\r\n      </g>\r\n     </g>\r\n    </g>\r\n    <g id=\"ytick_10\">\r\n     <g id=\"line2d_22\">\r\n      <g>\r\n       <use style=\"stroke:#000000;stroke-width:0.8;\" x=\"21.948438\" xlink:href=\"#m083fbb6f0c\" y=\"179.34\"/>\r\n      </g>\r\n     </g>\r\n     <g id=\"text_22\">\r\n      <!-- T -->\r\n      <g transform=\"translate(8.840625 183.139219)scale(0.1 -0.1)\">\r\n       <use xlink:href=\"#DejaVuSans-54\"/>\r\n      </g>\r\n     </g>\r\n    </g>\r\n    <g id=\"ytick_11\">\r\n     <g id=\"line2d_23\">\r\n      <g>\r\n       <use style=\"stroke:#000000;stroke-width:0.8;\" x=\"21.948438\" xlink:href=\"#m083fbb6f0c\" y=\"197.46\"/>\r\n      </g>\r\n     </g>\r\n     <g id=\"text_23\">\r\n      <!-- A -->\r\n      <g transform=\"translate(8.107813 201.259219)scale(0.1 -0.1)\">\r\n       <use xlink:href=\"#DejaVuSans-41\"/>\r\n      </g>\r\n     </g>\r\n    </g>\r\n    <g id=\"ytick_12\">\r\n     <g id=\"line2d_24\">\r\n      <g>\r\n       <use style=\"stroke:#000000;stroke-width:0.8;\" x=\"21.948438\" xlink:href=\"#m083fbb6f0c\" y=\"215.58\"/>\r\n      </g>\r\n     </g>\r\n     <g id=\"text_24\">\r\n      <!-- A -->\r\n      <g transform=\"translate(8.107813 219.379219)scale(0.1 -0.1)\">\r\n       <use xlink:href=\"#DejaVuSans-41\"/>\r\n      </g>\r\n     </g>\r\n    </g>\r\n   </g>\r\n   <g id=\"patch_3\">\r\n    <path d=\"M 21.948438 224.64 \r\nL 21.948438 7.2 \r\n\" style=\"fill:none;stroke:#000000;stroke-linecap:square;stroke-linejoin:miter;stroke-width:0.8;\"/>\r\n   </g>\r\n   <g id=\"patch_4\">\r\n    <path d=\"M 239.388438 224.64 \r\nL 239.388438 7.2 \r\n\" style=\"fill:none;stroke:#000000;stroke-linecap:square;stroke-linejoin:miter;stroke-width:0.8;\"/>\r\n   </g>\r\n   <g id=\"patch_5\">\r\n    <path d=\"M 21.948438 224.64 \r\nL 239.388437 224.64 \r\n\" style=\"fill:none;stroke:#000000;stroke-linecap:square;stroke-linejoin:miter;stroke-width:0.8;\"/>\r\n   </g>\r\n   <g id=\"patch_6\">\r\n    <path d=\"M 21.948438 7.2 \r\nL 239.388437 7.2 \r\n\" style=\"fill:none;stroke:#000000;stroke-linecap:square;stroke-linejoin:miter;stroke-width:0.8;\"/>\r\n   </g>\r\n  </g>\r\n </g>\r\n <defs>\r\n  <clipPath id=\"p0de5f4141f\">\r\n   <rect height=\"217.44\" width=\"217.44\" x=\"21.948438\" y=\"7.2\"/>\r\n  </clipPath>\r\n </defs>\r\n</svg>\r\n",
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAPYAAAD4CAYAAADfGBu4AAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjQuMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8+yak3AAAACXBIWXMAAAsTAAALEwEAmpwYAAALnElEQVR4nO3dQaiVZR7H8d9vnFKcaKGJMiRGCwkqlZTbJgk1cJkyEt51IBMxu4m2d1ZtXBWBuJMg7izCVpELXVQgmYZ34G6EEokpwbktwpG8Iv9Z3HNnLnLtPOe8z+N573++H5B73/ccH//Pfe+v5+2c9/0fR4QA5PK7SRcAoD6CDSREsIGECDaQEMEGEvp9q4Ef9/rYoD90HmfnrjsVqqnn2j82Vhurb3PDcDWPf1e/6t9ajLte7TG3ervrSW+Kl32o8zjnfrzavZiKDv9xT7Wx+jY3DFfz+Hf1dZzXL/HzqsHmVBxIiGADCRFsICGCDSRUHGzbR2yH7edaFgSgu1FW7GlJXw2+AuixomDbfkLSK5LelHS8aUUAOitdsV+X9HlEXJO0YHvvak+yfcL2ZduX7+lutSIBjKY02NOSZgffz+ohp+MRcToi9kXEvse0vkZ9AMYw9JJS25skHZT0ou2QtE5S2H4n6NIA9FLJin1M0kcRsSMinomI7ZKuS9rftjQA4yoJ9rSksw/s+0S8Og701tBT8Yg4sMq+99uUA6AGrjwDEiLYQEIEG0ioWQeVnbvu6Ny5q53H6dON7VLd5gi15larpr7VU0vW5hhThx/egYcVG0iIYAMJEWwgIYINJESwgYSKXhW3vVnS+cHmNkn3Jd0abE9FxGKD2gCMqSjYEbEgaY8k2Z6RdDsiTrYrC0AXnIoDCRFsIKGqwV7ZGunWwv2aQwMYQdVgr2yNtGXzuppDAxgBp+JAQgQbSGjku7siYqZBHQAqYsUGEiLYQEIEG0ioWQeVWvrUsaI2Op/8tqzzehRYsYGECDaQEMEGEiLYQEIEG0ioKNi2t9r+2Pb3tq/Yvmj7aOviAIxnaLBtW9Knkr6IiGcjYq+k45KeblwbgDGVvI99UNJiRJxa3hERNyR90KwqAJ2UnIo/L+nb1oUAqGfkF89sf2h7zvY3qzxGBxWgB0qCPS/ppeWNiHhb0iFJWx58Ih1UgH4oCfYFSRtsv7Vi38ZG9QCoYGiwIyIkHZH0qu3rti9JOiPp3ca1ARhT6QcG/KSlt7gArAFceQYkRLCBhAg2kJCXXhur70lvipd9qPM4fet+Uaurh9S/uWG4mse/q6/jvH6Jn73aY6zYQEIEG0iIYAMJEWwgIYINJESwgYSKg237iO2w/VzLggB0N8qKPS3pq8FXAD1W2szwCUmvSHpT3AwC9F7piv26pM8j4pqkBdt7V3vSyg4q93S3WpEARlMa7GlJs4PvZ/WQ0/GVHVQe0/oa9QEYw9D7sW1v0lKn0hdth6R1ksL2O9HqQnMAnZSs2MckfRQROyLimYjYLum6pP1tSwMwrpJgT0s6+8C+T8Sr40BvDT0Vj4gDq+x7v005AGrgyjMgIYINJESwgYSK2g+PY+euOzp37mrncfrUikaq286o1txq1dS3emrJ2s5q6vCdhz7Gig0kRLCBhAg2kBDBBhIi2EBCRa+K294s6fxgc5uk+5JuDbanImKxQW0AxlT6aZsLkvZIku0ZSbcj4mS7sgB0wak4kFDVYK/soHJr4X7NoQGMoGqwV3ZQ2bJ5Xc2hAYyAU3EgIYINJESwgYRGvrsrImYa1AGgIlZsICGCDSREsIGE3Krn/77dG+LSue1Nxka/Ze3E0jdTh3/Q5blfvdpjrNhAQgQbSIhgAwkRbCAhgg0kVBxs29tsz9r+zvYV25/Z3tmyOADjKW2NZC194uaZiDg+2Ldb0lZJ19qVB2AcpdeKH5B0LyJOLe+IiLk2JQHoqvRU/AVJV4Y9iQ4qQD/QQQVIqDTY85L2tiwEQD2lwb4gab3tE8s7bO+yvb9NWQC6KAp2LN0pclTSa4O3u+YlvSfpZsviAIynuINKRPwo6Y2GtQCohCvPgIQINpAQwQYSGrlL6aNWqxtHLTW7evSt00jf6qml5u9Q3+b2MKzYQEIEG0iIYAMJEWwgIYINJFTaaGGzpPODzW2S7ku6NdieiojFBrUBGFNRsCNiQdIeSbI9I+l2RJxsVxaALjgVBxKqGmw6qAD9QAcVICFOxYGECDaQEMEGEhr57q6ImGlQB4CKWLGBhAg2kBDBBhLyUmfh+vbt3hCXzm1vMjb6LWsnlr6ZOvyDLs/96tUeY8UGEiLYQEIEG0iIYAMJEWwgoeJg295me3bwoXxXbH9me2fL4gCMp7Q1kiWdlXQmIo4P9u2WtFXStXblARhH6bXiByTdi4hTyzsiYq5NSQC6Kj0Vf0HSlWFPooMK0A90UAESKg32vKS9LQsBUE9psC9IWm/7xPIO27ts729TFoAuioIdS3eKHJX02uDtrnlJ70m62bI4AOMp7qASET9KeqNhLQAq4cozICGCDSREsIGERu5Siv7pW8eSWuP0bV5rCSs2kBDBBhIi2EBCBBtIiGADCRUF2/ZW2x/b/n7QPeWi7aOtiwMwnqHBHnRP+VTSFxHxbETslXRc0tONawMwppL3sQ9KWnyge8oNSR80qwpAJyWn4s9L+rZkMDqoAP0w8otntj+0PWf7mwcfo4MK0A8lwZ6X9NLyRkS8LemQpC2tigLQTUmwL0jaYPutFfs2NqoHQAVDgz3onnJE0qu2r9u+JOmMpHcb1wZgTEV3d0XET1p6iwvAGsCVZ0BCBBtIiGADCfW+g0qtLhq11OzG0bcOIX2rp2/zktZONxZWbCAhgg0kRLCBhAg2kBDBBhIqelXc9mZJ5web2yTdl3RrsD0VEYsNagMwptJLShck7ZEk2zOSbkfEyXZlAeiCU3EgIYINJFQ12LRGAvqharBpjQT0A6fiQEIEG0ho5Lu7ImKmQR0AKmLFBhIi2EBCBBtIiGADCXmpbXh9T3pTvOxDncfpWyua/8c2O/ifPrXq+jrO65f42as9xooNJESwgYQINpAQwQYSKg627SO2w/ZzLQsC0N0oK/a0pK8GXwH0WFGwbT8h6RVJb4pP3QR6r3TFfl3S5xFxTdKC7b0NawLQUWmwpyXNDr6f1UNOx1d2ULmnuzXqAzCGobdt2t4k6aCkF22HpHWSwvY78cBlaxFxWtJpaenKswb1AihQsmIfk/RRROyIiGciYruk65L2ty0NwLhKgj0t6ewD+z4Rr44DvTX0VDwiDqyy7/025QCogSvPgIQINpAQwQYSIthAQgQbSIhgAwkRbCAhgg0kRLCBhAg2kBCtkYCEaI0EJERrJCChqq2R6KAC9EPV1kgRcToi9kXEvse0vkZ9AMZQtTUSgH6gNRKQEK2RgIRojQQkxJVnQEIEG0iIYAMJudU7VrZvSbox5GlPSfpXhX+OcR7dWIzTn3F2RMSW1R5oFuwSti9HxD7GaTtOH2tinLbjcCoOJESwgYQmHezTjPNIxqk5FuOsgXEm+v/YANqY9IoNoAGCDSQ0sWB37aFme7Ptq4M/N23/c8X24yOOtdX2x7a/t33F9kXbR8esqzfzGoy3zfas7e8Gc/vM9s5J1VSjnsE4nY9Zg591lb6AVcaJiIn8kfR3SV9K+luFsWYk/XXMv2tJFyX9ecW+HZL+spbn9Rtz2y1pf49+1iPXU/uY1fhZ1zz2NcaZyIrdsx5qByUtRsSp5R0RcSMiPhh1oJ7NS5IOSLr3wNzmIuLLNV5PtWNWS61jX2ucSZ2KF/VQe0Sel/RtpbH6NC9JekHSlQnXsFKtemoes1pqHfsq40wq2EU91CbB9oe252x/M8Zf7+28Mut4zGqpdeyrjDO00UJtPeyhNi/pT8sbEfG27ackXR5lkB7OS1qa27EJ/durqVVPlWNWS61jX/N3aBIrdt96qF2QtMH2Wyv2bRxjnL7NS1qa23rbJ5Z32N5le5I/6xr11DpmtdQ69tV+hyYR7F71UBv8l/CIpFdtX7d9SdIZSe+OOFSv5iX9d25HJb02eHtpXtJ7km6u5XoqHrNaah37ar9DXFIKJMSVZ0BCBBtIiGADCRFsICGCDSREsIGECDaQ0H8ANcuvUT2TA3cAAAAASUVORK5CYII=\n"
     },
     "metadata": {
      "needs_background": "light"
     }
    }
   ],
   "source": [
    "#Using dotplot on DNA\n",
    "\n",
    "dna1 = Seq(\"ATGATCTCGTAA\")\n",
    "dna2 = Seq(\"ATTATGTCGTAA\")\n",
    "\n",
    "dotplotx(dna1,dna2)"
   ]
  },
  {
   "source": [
    "# 8) Working with Biological Data File Formats"
   ],
   "cell_type": "markdown",
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO"
   ]
  },
  {
   "source": [
    "## FASTA"
   ],
   "cell_type": "markdown",
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "MN908947.3\nMN908947.3 Severe acute respiratory syndrome coronavirus 2 isolate Wuhan-Hu-1, complete genome\n"
     ]
    }
   ],
   "source": [
    "#Load FASTA format of SARS COVID 19 GENOME\n",
    "for record in SeqIO.parse(\"Sars Covid Sequence.fasta\",\"fasta\"):\n",
    "    print(record.id)\n",
    "    print(record.description)\n",
    "    #print(record.seq) - very long "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "Seq('ATTAAAGGTTTATACCTTCCCAGGTAACAAACCAACCAACTTTCGATCTCTTGT...AAA')"
      ]
     },
     "metadata": {},
     "execution_count": 11
    }
   ],
   "source": [
    "#Read FASTA condensed sequence\n",
    "fasta_dna_record = SeqIO.read(\"SARS covid sequence.fasta\",\"fasta\")\n",
    "dna_sequence = fasta_dna_record.seq\n",
    "dna_sequence"
   ]
  },
  {
   "source": [
    "## Genbank"
   ],
   "cell_type": "markdown",
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "MN908947.3\nSevere acute respiratory syndrome coronavirus 2 isolate Wuhan-Hu-1, complete genome\n"
     ]
    }
   ],
   "source": [
    "#Load Genbank format of SARS COVID 19 GENOME\n",
    "for record in SeqIO.parse(\"SARS covid sequence.gb\",\"genbank\"):\n",
    "    print(record.id)\n",
    "    print(record.description)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "Seq('ATTAAAGGTTTATACCTTCCCAGGTAACAAACCAACCAACTTTCGATCTCTTGT...AAA')"
      ]
     },
     "metadata": {},
     "execution_count": 20
    }
   ],
   "source": [
    "#Read Genbank condensed sequence\n",
    "gb_dna_record = SeqIO.read(\"SARS covid sequence.fasta\",\"fasta\")\n",
    "dna_sequence = gb_dna_record.seq\n",
    "dna_sequence"
   ]
  },
  {
   "source": [
    "# 9) Working with 3D structures"
   ],
   "cell_type": "markdown",
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.PDB import PDBParser, MMCIFParser"
   ]
  },
  {
   "source": [
    "## PDB Files"
   ],
   "cell_type": "markdown",
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "C:\\Users\\18687\\AppData\\Local\\Programs\\Python\\Python39\\lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 2837.\n  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "# Creating Parser\n",
    "parser = PDBParser()\n",
    "structure = parser.get_structure(\"SARS covid sequence\",\"SARS covid sequence.pdb\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "<Structure id=SARS covid sequence>"
      ]
     },
     "metadata": {},
     "execution_count": 26
    }
   ],
   "source": [
    "structure #id = <Structure id=SARS covid sequence>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "#No. Models in the structure = 1\r\n",
    "len(structure)\r\n",
    "\r\n",
    "#Model\r\n",
    "model = structure[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "Chain <Chain id=A>, Chain_ID A\nChain <Chain id=C>, Chain_ID C\n"
     ]
    }
   ],
   "source": [
    "#FORMAT: Structure -> Model -> Chain ->Residue -> Atom\n",
    "\n",
    "#Check for chains\n",
    "for chain in model:\n",
    "    print(f\"Chain {chain}, Chain_ID {chain.id}\")\n",
    "\n",
    "#Check for residue\n",
    "for chain in model:\n",
    "    for residue in chain:\n",
    "        pass\n",
    "        #print(residue) - a lot \n",
    "\n",
    "#Check for atoms\n",
    "for chain in model:\n",
    "    for residue in chain:\n",
    "        for atom in residue:\n",
    "            pass\n",
    "            #print(atom) - a lot\n",
    "        "
   ]
  },
  {
   "source": [
    "# 10) Visualizing 3D structures"
   ],
   "cell_type": "markdown",
   "metadata": {}
  },
  {
   "source": [
    "## Py3Dmol"
   ],
   "cell_type": "markdown",
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "import py3Dmol "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "output_type": "display_data",
     "data": {
      "application/3dmoljs_load.v0": "<div id=\"3dmolviewer_16186879641375794\"  style=\"position: relative; width: 640px; height: 480px\">\n        <p id=\"3dmolwarning_16186879641375794\" style=\"background-color:#ffcccc;color:black\">You appear to be running in JupyterLab (or JavaScript failed to load for some other reason).  You need to install the 3dmol extension: <br>\n        <tt>jupyter labextension install jupyterlab_3dmol</tt></p>\n        </div>\n<script>\n\nvar loadScriptAsync = function(uri){\n  return new Promise((resolve, reject) => {\n    var tag = document.createElement('script');\n    tag.src = uri;\n    tag.async = true;\n    tag.onload = () => {\n      resolve();\n    };\n  var firstScriptTag = document.getElementsByTagName('script')[0];\n  firstScriptTag.parentNode.insertBefore(tag, firstScriptTag);\n});\n};\n\nif(typeof $3Dmolpromise === 'undefined') {\n$3Dmolpromise = null;\n  $3Dmolpromise = loadScriptAsync('https://3dmol.org/build/3Dmol.js');\n}\n\nvar viewer_16186879641375794 = null;\nvar warn = document.getElementById(\"3dmolwarning_16186879641375794\");\nif(warn) {\n    warn.parentNode.removeChild(warn);\n}\n$3Dmolpromise.then(function() {\nviewer_16186879641375794 = $3Dmol.createViewer($(\"#3dmolviewer_16186879641375794\"),{backgroundColor:\"white\"});\n$3Dmol.download(\"pdb:6LU7\", viewer_16186879641375794, {}, function() {\nviewer_16186879641375794.zoomTo();\n\tviewer_16186879641375794.setStyle({\"cartoon\": {\"color\": \"spectrum\"}});\nviewer_16186879641375794.render();\n})\n});\n</script>",
      "text/html": "<div id=\"3dmolviewer_16186879641375794\"  style=\"position: relative; width: 640px; height: 480px\">\n        <p id=\"3dmolwarning_16186879641375794\" style=\"background-color:#ffcccc;color:black\">You appear to be running in JupyterLab (or JavaScript failed to load for some other reason).  You need to install the 3dmol extension: <br>\n        <tt>jupyter labextension install jupyterlab_3dmol</tt></p>\n        </div>\n<script>\n\nvar loadScriptAsync = function(uri){\n  return new Promise((resolve, reject) => {\n    var tag = document.createElement('script');\n    tag.src = uri;\n    tag.async = true;\n    tag.onload = () => {\n      resolve();\n    };\n  var firstScriptTag = document.getElementsByTagName('script')[0];\n  firstScriptTag.parentNode.insertBefore(tag, firstScriptTag);\n});\n};\n\nif(typeof $3Dmolpromise === 'undefined') {\n$3Dmolpromise = null;\n  $3Dmolpromise = loadScriptAsync('https://3dmol.org/build/3Dmol.js');\n}\n\nvar viewer_16186879641375794 = null;\nvar warn = document.getElementById(\"3dmolwarning_16186879641375794\");\nif(warn) {\n    warn.parentNode.removeChild(warn);\n}\n$3Dmolpromise.then(function() {\nviewer_16186879641375794 = $3Dmol.createViewer($(\"#3dmolviewer_16186879641375794\"),{backgroundColor:\"white\"});\n$3Dmol.download(\"pdb:6LU7\", viewer_16186879641375794, {}, function() {\nviewer_16186879641375794.zoomTo();\n\tviewer_16186879641375794.setStyle({\"cartoon\": {\"color\": \"spectrum\"}});\nviewer_16186879641375794.render();\n})\n});\n</script>"
     },
     "metadata": {}
    },
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "<py3Dmol.view at 0x19c8b266490>"
      ]
     },
     "metadata": {},
     "execution_count": 40
    }
   ],
   "source": [
    "#py3Dmol.view(query = \"type:ID\")\n",
    "view1 = py3Dmol.view(query = \"pdb:6LU7\") #Specify ID from website, not download\n",
    "\n",
    "view1.setStyle({'cartoon': {'color':'spectrum'}})"
   ]
  },
  {
   "source": [
    "## Nglview - shows atom location and tertiary structure"
   ],
   "cell_type": "markdown",
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Enabling notebook extension nglview-js-widgets/extension...\n      - Validating: ok\n"
     ]
    }
   ],
   "source": [
    "import nglview as nv\n",
    "!jupyter-nbextension enable nglview --py --sys-prefix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "#nv.demo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "output_type": "display_data",
     "data": {
      "text/plain": "NGLWidget()",
      "application/vnd.jupyter.widget-view+json": {
       "version_major": 2,
       "version_minor": 0,
       "model_id": "7531604d1df24b4194aa7c527fa12ecc"
      }
     },
     "metadata": {}
    }
   ],
   "source": [
    "view2 = nv.show_biopython(structure)\n",
    "view2\n",
    "\n",
    "#Static image\n",
    "#view2.render_image()"
   ]
  },
  {
   "source": [
    "## pytraj"
   ],
   "cell_type": "markdown",
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "#import pytraj as pt\n",
    "# ncov_traj = pt.load(\"6LU7.pdb\")\n",
    "# view3 = nv.show_pytraj(ncov_traj)\n",
    "# view3"
   ]
  },
  {
   "source": [
    "# 11) BLAST\n",
    "can be used online or locally with a blast CLI or BioPython"
   ],
   "cell_type": "markdown",
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Blast import NCBIWWW"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "#NOT RUN because the result will be LOST\n",
    "\n",
    "# result_handle = NCBIWWW.qblast(\"blastn\",\"nt\",\"SARS covid sequence.fasta\")\n",
    "\n",
    "# #result_handle is a generator so we have to loop through it to obtain results\n",
    "# for i in result_handle:\n",
    "#     print(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Best option working with BLAST \n",
    "\n",
    "with NCBIWWW.qblast(\"blastn\",\"nt\",\"SARS covid sequence.fasta\") as result_handle:\n",
    "    with open(r\"result_blast_covid.xml\",\"w\") as xml_file:\n",
    "        xml_file.write(result_handle)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ]
}